// react-lightning-design-system/module/scripts/BreadCrumbs.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import classnames from 'classnames';
export var Crumb = function Crumb(_ref) {
  var className = _ref.className,
      href = _ref.href,
      children = _ref.children,
      props = _objectWithoutProperties(_ref, ["className", "href", "children"]);

  var text = children;
  var cClassName = classnames('slds-list__item slds-text-heading_label', className);
  return /*#__PURE__*/React.createElement("li", _extends({}, props, {
    className: cClassName
  }), /*#__PURE__*/React.createElement("a", {
    href: href
  }, text));
};
export var BreadCrumbs = function BreadCrumbs(_ref2) {
  var label = _ref2.label,
      className = _ref2.className,
      children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["label", "className", "children"]);

  var oClassName = classnames('slds-breadcrumb slds-list_horizontal', className);
  return /*#__PURE__*/React.createElement("nav", _extends({}, props, {
    role: "navigation"
  }), label ? /*#__PURE__*/React.createElement("p", {
    id: "bread-crumb-label",
    className: "slds-assistive-text"
  }, label) : null, /*#__PURE__*/React.createElement("ol", {
    className: oClassName,
    "aria-labelledby": "bread-crumb-label"
  }, children));
};
