// react-lightning-design-system/module/scripts/Datepicker.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classnames from 'classnames';
import moment from 'moment';
import { Button } from './Button';
import { Select, Option } from './Select';
import { getToday, isElInChildren } from './util';

function createCalendarObject(date, mnDate, mxDate) {
  var minDate;
  var maxDate;
  var d = moment(date, 'YYYY-MM-DD');

  if (!d.isValid()) {
    d = moment(getToday(), 'YYYY-MM-DD');
  }

  if (mnDate) {
    var minD = moment(mnDate, 'YYYY-MM-DD');

    if (minD.isValid()) {
      minDate = {
        year: minD.year(),
        month: minD.month(),
        date: minD.date(),
        value: minD.format('YYYY-MM-DD')
      };
    }
  }

  if (mxDate) {
    var maxD = moment(mxDate, 'YYYY-MM-DD');

    if (maxD.isValid()) {
      maxDate = {
        year: maxD.year(),
        month: maxD.month(),
        date: maxD.date(),
        value: maxD.format('YYYY-MM-DD')
      };
    }
  }

  var year = d.year();
  var month = d.month();
  var first = moment(d).startOf('month').startOf('week');
  var last = moment(d).endOf('month').endOf('week');
  var weeks = [];
  var days = [];

  for (var dd = first; dd.isBefore(last); dd = dd.add(1, 'd')) {
    days.push({
      year: dd.year(),
      month: dd.month(),
      date: dd.date(),
      value: dd.format('YYYY-MM-DD')
    });

    if (days.length === 7) {
      weeks.push(days);
      days = [];
    }
  }

  var cal = {
    year: year,
    month: month,
    weeks: weeks
  };

  if (minDate) {
    cal.minDate = minDate;
  }

  if (maxDate) {
    cal.maxDate = maxDate;
  }

  return cal;
}

function cancelEvent(e) {
  e.preventDefault();
  e.stopPropagation();
}

export var Datepicker = /*#__PURE__*/function (_Component) {
  _inherits(Datepicker, _Component);

  var _super = _createSuper(Datepicker);

  function Datepicker(props) {
    var _this;

    _classCallCheck(this, Datepicker);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "node", null);

    _defineProperty(_assertThisInitialized(_this), "month", null);

    _this.state = {};
    _this.onBlur = _this.onBlur.bind(_assertThisInitialized(_this));
    _this.onKeyDown = _this.onKeyDown.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(Datepicker, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      if (this.props.autoFocus) {
        var targetDate = this.props.selectedDate || getToday();
        setTimeout(function () {
          _this2.focusDate(targetDate);
        }, 10);
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (this.state.focusDate && (this.state.targetDate || this.props.selectedDate)) {
        this.focusDate(this.state.targetDate || this.props.selectedDate);
        /* eslint-disable react/no-did-update-set-state */

        this.setState({
          focusDate: false
        });
      }
    }
  }, {
    key: "onDateKeyDown",
    value: function onDateKeyDown(date, e) {
      var targetDate = this.state.targetDate || this.props.selectedDate;

      if (e.keyCode === 13 || e.keyCode === 32) {
        // return / space
        this.onDateClick(date);
        e.preventDefault();
        e.stopPropagation();
      } else if (e.keyCode >= 37 && e.keyCode <= 40) {
        // cursor key
        if (e.keyCode === 37) {
          targetDate = moment(targetDate).add(-1, e.shiftKey ? 'months' : 'days');
        } else if (e.keyCode === 39) {
          // right arrow key
          targetDate = moment(targetDate).add(1, e.shiftKey ? 'months' : 'days');
        } else if (e.keyCode === 38) {
          // up arrow key
          targetDate = moment(targetDate).add(-1, e.shiftKey ? 'years' : 'weeks');
        } else if (e.keyCode === 40) {
          // down arrow key
          targetDate = moment(targetDate).add(1, e.shiftKey ? 'years' : 'weeks');
        }

        targetDate = targetDate.format('YYYY-MM-DD');
        this.setState({
          targetDate: targetDate,
          focusDate: true
        });
        e.preventDefault();
        e.stopPropagation();
      }
    }
  }, {
    key: "onDateClick",
    value: function onDateClick(date) {
      if (this.props.onSelect) {
        this.props.onSelect(date);
      }
    }
  }, {
    key: "onDateFocus",
    value: function onDateFocus(date) {
      var _this3 = this;

      if (this.state.targetDate !== date) {
        setTimeout(function () {
          _this3.setState({
            targetDate: date
          });
        }, 10);
      }
    }
  }, {
    key: "onYearChange",
    value: function onYearChange(e) {
      // eslint-disable-next-line react/no-access-state-in-setstate
      var targetDate = this.state.targetDate || this.props.selectedDate;
      targetDate = moment(targetDate).year(Number(e.target.value)).format('YYYY-MM-DD');
      this.setState({
        targetDate: targetDate
      });
    }
  }, {
    key: "onMonthChange",
    value: function onMonthChange(month) {
      // eslint-disable-next-line react/no-access-state-in-setstate
      var targetDate = this.state.targetDate || this.props.selectedDate;
      targetDate = moment(targetDate).add(month, 'months').format('YYYY-MM-DD');
      this.setState({
        targetDate: targetDate
      });
    }
  }, {
    key: "onBlur",
    value: function onBlur(e) {
      var _this4 = this;

      setTimeout(function () {
        if (!_this4.isFocusedInComponent()) {
          if (_this4.props.onBlur) {
            _this4.props.onBlur(e);
          }
        }
      }, 10);
    }
  }, {
    key: "onKeyDown",
    value: function onKeyDown(e) {
      if (e.keyCode === 27) {
        // ESC
        if (this.props.onClose) {
          this.props.onClose();
        }
      }
    }
  }, {
    key: "focusDate",
    value: function focusDate(date) {
      var el = this.month;

      if (!el) {
        return;
      }

      var dateEl = el.querySelector(".slds-day[data-date-value=\"".concat(date, "\"]"));

      if (dateEl) {
        dateEl.focus();
      }
    }
  }, {
    key: "isFocusedInComponent",
    value: function isFocusedInComponent() {
      return isElInChildren(this.node, document.activeElement);
    }
  }, {
    key: "renderFilter",
    value: function renderFilter(cal) {
      return /*#__PURE__*/React.createElement("div", {
        className: "slds-datepicker__filter slds-grid"
      }, /*#__PURE__*/React.createElement("div", {
        className: "slds-datepicker__filter_month slds-grid slds-grid_align-spread slds-size_2-of-3"
      }, /*#__PURE__*/React.createElement("div", {
        className: "slds-align-middle"
      }, /*#__PURE__*/React.createElement(Button, {
        className: "slds-align-middle",
        type: "icon-container",
        icon: "left",
        size: "small",
        alt: "Previous Month",
        onClick: this.onMonthChange.bind(this, -1)
      })), /*#__PURE__*/React.createElement("h2", {
        className: "slds-align-middle"
      }, moment.monthsShort()[cal.month]), /*#__PURE__*/React.createElement("div", {
        className: "slds-align-middle"
      }, /*#__PURE__*/React.createElement(Button, {
        className: "slds-align-middle",
        type: "icon-container",
        icon: "right",
        size: "small",
        alt: "Next Month",
        onClick: this.onMonthChange.bind(this, 1)
      }))), /*#__PURE__*/React.createElement("div", {
        className: "slds-size_1-of-3"
      }, /*#__PURE__*/React.createElement(Select, {
        value: cal.year,
        onChange: this.onYearChange.bind(this)
      }, new Array(11).join('_').split('_').map(function (a, i) {
        var year = cal.year + i - 5;
        return /*#__PURE__*/React.createElement(Option, {
          key: year,
          label: String(year),
          value: year
        });
      }))));
    }
  }, {
    key: "renderMonth",
    value: function renderMonth(cal, selectedDate, today) {
      var _this5 = this;

      return /*#__PURE__*/React.createElement("table", {
        className: "datepicker__month",
        role: "grid",
        "aria-labelledby": "month",
        ref: function ref(node) {
          return _this5.month = node;
        }
      }, /*#__PURE__*/React.createElement("thead", null, /*#__PURE__*/React.createElement("tr", null, moment.weekdaysMin(true).map(function (wd, i) {
        return (
          /*#__PURE__*/
          // eslint-disable-next-line react/no-array-index-key
          React.createElement("th", {
            key: i
          }, /*#__PURE__*/React.createElement("abbr", {
            title: moment.weekdays(true, i)
          }, wd))
        );
      }))), /*#__PURE__*/React.createElement("tbody", null, cal.weeks.map(function (days, i) {
        return (
          /*#__PURE__*/
          // eslint-disable-next-line react/no-array-index-key
          React.createElement("tr", {
            key: i
          }, days.map(_this5.renderDate.bind(_this5, cal, selectedDate, today)))
        );
      })));
    }
  }, {
    key: "renderDate",
    value: function renderDate(cal, selectedDate, today, d, i) {
      var selectable = true;
      var enabled = d.year === cal.year && d.month === cal.month;

      if (cal.minDate) {
        var min = moment(d.value, 'YYYY-MM-DD').isAfter(moment(cal.minDate.value, 'YYYY-MM-DD'));
        selectable = selectable && min;
        enabled = enabled && min;
      }

      if (cal.maxDate) {
        var max = moment(d.value, 'YYYY-MM-DD').isBefore(moment(cal.maxDate.value, 'YYYY-MM-DD'));
        selectable = selectable && max;
        enabled = enabled && max;
      }

      var selected = d.value === selectedDate;
      var isToday = d.value === today;
      var dateClassName = classnames({
        'slds-disabled-text': !enabled,
        'slds-is-selected': selected,
        'slds-is-today': isToday
      });
      return /*#__PURE__*/React.createElement("td", {
        className: dateClassName,
        key: i,
        headers: moment.weekdays(i),
        role: "gridcell",
        "aria-disabled": !enabled,
        "aria-selected": selected
      }, /*#__PURE__*/React.createElement("span", {
        className: "slds-day" // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        ,
        tabIndex: selectable ? 0 : -1,
        onClick: selectable ? this.onDateClick.bind(this, d.value) : undefined,
        onKeyDown: selectable ? this.onDateKeyDown.bind(this, d.value) : undefined,
        onFocus: enabled ? this.onDateFocus.bind(this, d.value) : cancelEvent,
        "data-date-value": d.value
      }, d.date));
    }
  }, {
    key: "render",
    value: function render() {
      var _this6 = this;

      var _this$props = this.props,
          className = _this$props.className,
          selectedDate = _this$props.selectedDate,
          minDate = _this$props.minDate,
          maxDate = _this$props.maxDate,
          elementRef = _this$props.elementRef,
          ExtensionRenderer = _this$props.extensionRenderer,
          autoFocus = _this$props.autoFocus,
          onSelect = _this$props.onSelect,
          onClose = _this$props.onClose,
          props = _objectWithoutProperties(_this$props, ["className", "selectedDate", "minDate", "maxDate", "elementRef", "extensionRenderer", "autoFocus", "onSelect", "onClose"]);

      var today = getToday();
      var targetDate = this.state.targetDate || selectedDate;
      var cal = createCalendarObject(targetDate, minDate, maxDate);
      var datepickerClassNames = classnames('slds-datepicker', className);

      var handleDOMRef = function handleDOMRef(node) {
        _this6.node = node;

        if (elementRef) {
          elementRef(node);
        }
      };

      return /*#__PURE__*/React.createElement("div", _extends({}, props, {
        className: datepickerClassNames,
        ref: handleDOMRef,
        tabIndex: -1,
        "aria-hidden": false,
        onBlur: this.onBlur,
        onKeyDown: this.onKeyDown
      }), this.renderFilter(cal), this.renderMonth(cal, selectedDate, today), ExtensionRenderer ? /*#__PURE__*/React.createElement(ExtensionRenderer, this.props) : undefined);
    }
  }]);

  return Datepicker;
}(Component);
