// react-lightning-design-system/module/scripts/DropdownButton.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classnames from 'classnames';
import { Button } from './Button';
import { DropdownMenu } from './DropdownMenu';
import { registerStyle, isElInChildren } from './util';
export var DropdownButton = /*#__PURE__*/function (_Component) {
  _inherits(DropdownButton, _Component);

  var _super = _createSuper(DropdownButton);

  function DropdownButton(props) {
    var _this;

    _classCallCheck(this, DropdownButton);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "node", null);

    _defineProperty(_assertThisInitialized(_this), "trigger", null);

    _defineProperty(_assertThisInitialized(_this), "dropdown", null);

    _defineProperty(_assertThisInitialized(_this), "onBlur", function () {
      setTimeout(function () {
        if (!_this.isFocusedInComponent()) {
          _this.setState({
            opened: false
          });

          if (_this.props.onBlur) {
            _this.props.onBlur();
          }
        }
      }, 10);
    });

    _defineProperty(_assertThisInitialized(_this), "onKeyDown", function (e) {
      if (e.keyCode === 40) {
        // down
        e.preventDefault();
        e.stopPropagation();

        if (!_this.state.opened) {
          _this.setState({
            opened: true
          });

          if (_this.props.onClick) {
            _this.props.onClick(e);
          }

          setTimeout(function () {
            _this.focusToTargetItemEl();
          }, 20);
        } else {
          _this.focusToTargetItemEl();
        }
      } else if (e.keyCode === 27) {
        // ESC
        e.preventDefault();
        e.stopPropagation();

        _this.setState({
          opened: false
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onTriggerClick", function (e) {
      if (!_this.props.hoverPopup) {
        _this.setState(function (prevState) {
          return {
            opened: !prevState.opened
          };
        });
      }

      if (_this.props.onClick) {
        _this.props.onClick(e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onMenuSelect", function (eventKey) {
      if (!_this.props.hoverPopup) {
        setTimeout(function () {
          var triggerElem = _this.trigger;
          if (triggerElem) triggerElem.focus();

          _this.setState({
            opened: false
          });
        }, 10);
      }

      if (_this.props.onMenuSelect) {
        _this.props.onMenuSelect(eventKey);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onMenuClose", function () {
      if (_this.trigger) {
        _this.trigger.focus();
      }

      _this.setState({
        opened: false
      });
    });

    _this.state = {
      opened: false
    };
    registerStyle('no-hover-popup', [['.slds-dropdown-trigger:hover .slds-dropdown_menu.react-slds-no-hover-popup', '{ visibility: hidden; opacity: 0; }'], ['.slds-dropdown-trigger.react-slds-dropdown-opened .slds-dropdown_menu', '{ visibility: visible !important; opacity: 1 !important; }']]);
    return _this;
  }

  _createClass(DropdownButton, [{
    key: "isFocusedInComponent",
    value: function isFocusedInComponent() {
      var targetEl = document.activeElement;
      return isElInChildren(this.node, targetEl) || isElInChildren(this.dropdown, targetEl);
    }
  }, {
    key: "focusToTargetItemEl",
    value: function focusToTargetItemEl() {
      var dropdownEl = this.dropdown;

      if (!dropdownEl) {
        return;
      }

      var firstItemEl = dropdownEl.querySelector('.slds-is-selected > .react-slds-menuitem[tabIndex]') || dropdownEl.querySelector('.react-slds-menuitem[tabIndex]');

      if (firstItemEl) {
        firstItemEl.focus();
      }
    }
  }, {
    key: "renderButton",
    value: function renderButton(_ref) {
      var _this2 = this;

      var grouped = _ref.grouped,
          isFirstInGroup = _ref.isFirstInGroup,
          isLastInGroup = _ref.isLastInGroup,
          props = _objectWithoutProperties(_ref, ["grouped", "isFirstInGroup", "isLastInGroup"]);

      var pprops = props;
      delete pprops.onMenuItemClick;
      var button = /*#__PURE__*/React.createElement(Button, _extends({}, pprops, {
        "aria-haspopup": true,
        buttonRef: function buttonRef(node) {
          return _this2.trigger = node;
        },
        onClick: this.onTriggerClick,
        onKeyDown: this.onKeyDown,
        onBlur: this.onBlur
      }));

      if (grouped) {
        var noneStyle = {
          display: 'none'
        };
        return /*#__PURE__*/React.createElement("div", {
          className: "slds-button-group"
        }, isFirstInGroup ? null : /*#__PURE__*/React.createElement("button", {
          type: "button",
          className: "slds-button",
          style: noneStyle
        }), button, isLastInGroup ? null : /*#__PURE__*/React.createElement("button", {
          type: "button",
          className: "slds-button",
          style: noneStyle
        }));
      }

      return button;
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props = this.props,
          className = _this$props.className,
          menuAlign = _this$props.menuAlign,
          menuSize = _this$props.menuSize,
          nubbinTop = _this$props.nubbinTop,
          hoverPopup = _this$props.hoverPopup,
          menuHeader = _this$props.menuHeader,
          type = _this$props.type,
          label = _this$props.label,
          children = _this$props.children,
          style = _this$props.style,
          menuStyle = _this$props.menuStyle,
          props = _objectWithoutProperties(_this$props, ["className", "menuAlign", "menuSize", "nubbinTop", "hoverPopup", "menuHeader", "type", "label", "children", "style", "menuStyle"]);

      var icon = this.props.icon;
      var dropdownClassNames = classnames(className, 'slds-dropdown-trigger', {
        'slds-button-space-left': !props.grouped,
        'react-slds-dropdown-opened': this.state.opened
      });
      var iconMore = null;

      if (!label && !icon) {
        icon = 'down';
      }

      if (label || type === 'icon-more') {
        iconMore = 'down';
      }

      var dropdown = /*#__PURE__*/React.createElement(DropdownMenu, {
        portalClassName: className,
        align: menuAlign,
        header: menuHeader,
        size: menuSize,
        nubbinTop: nubbinTop,
        hoverPopup: hoverPopup,
        dropdownMenuRef: function dropdownMenuRef(node) {
          return _this3.dropdown = node;
        },
        onMenuSelect: this.onMenuSelect,
        onMenuClose: this.onMenuClose,
        onBlur: this.onBlur,
        style: Object.assign({
          transition: 'none'
        }, menuStyle)
      }, children);
      return /*#__PURE__*/React.createElement("div", {
        className: dropdownClassNames,
        style: style,
        ref: function ref(node) {
          return _this3.node = node;
        }
      }, this.renderButton(_objectSpread({
        type: type,
        label: label,
        icon: icon,
        iconMore: iconMore
      }, props)), hoverPopup || this.state.opened ? dropdown : undefined);
    }
  }]);

  return DropdownButton;
}(Component);
