// react-lightning-design-system/module/scripts/FormElement.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _typeof from "@babel/runtime/helpers/typeof";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React from 'react';
import classnames from 'classnames';
export var FormElement = /*#__PURE__*/function (_React$Component) {
  _inherits(FormElement, _React$Component);

  var _super = _createSuper(FormElement);

  function FormElement() {
    _classCallCheck(this, FormElement);

    return _super.apply(this, arguments);
  }

  _createClass(FormElement, [{
    key: "renderFormElement",
    value: function renderFormElement(props) {
      var className = props.className,
          error = props.error,
          totalCols = props.totalCols,
          _props$cols = props.cols,
          cols = _props$cols === void 0 ? 1 : _props$cols,
          formElementRef = props.formElementRef,
          children = props.children;
      var formElementClassNames = classnames('slds-form-element', _defineProperty({
        'slds-has-error': error
      }, "slds-size_".concat(cols, "-of-").concat(totalCols), typeof totalCols === 'number'), className);
      return /*#__PURE__*/React.createElement("div", {
        ref: formElementRef,
        key: "form-element",
        className: formElementClassNames
      }, children);
    }
  }, {
    key: "renderLabel",
    value: function renderLabel() {
      var _this$props = this.props,
          id = _this$props.id,
          label = _this$props.label,
          required = _this$props.required;
      return label ? /*#__PURE__*/React.createElement("label", {
        key: "form-element-label",
        className: "slds-form-element__label",
        htmlFor: id
      }, label, required ? /*#__PURE__*/React.createElement("abbr", {
        className: "slds-required"
      }, "*") : undefined) : undefined;
    }
  }, {
    key: "renderControl",
    value: function renderControl(props) {
      var children = props.children,
          dropdown = props.dropdown,
          error = props.error;
      var readOnly = this.props.readOnly;
      var formElementControlClassNames = classnames('slds-form-element__control', {
        'slds-has-divider_bottom': readOnly
      });
      return /*#__PURE__*/React.createElement("div", {
        key: "form-element-control",
        className: formElementControlClassNames
      }, children, dropdown, this.renderError(error));
    }
  }, {
    key: "renderError",
    value: function renderError(error) {
      var errorMessage = error ? typeof error === 'string' ? error : _typeof(error) === 'object' ? error.message : undefined : undefined;
      return errorMessage ? /*#__PURE__*/React.createElement("span", {
        key: "slds-form-error",
        className: "slds-form-element__help"
      }, errorMessage) : undefined;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          dropdown = _this$props2.dropdown,
          className = _this$props2.className,
          totalCols = _this$props2.totalCols,
          cols = _this$props2.cols,
          error = _this$props2.error,
          children = _this$props2.children,
          style = _this$props2.style,
          props = _objectWithoutProperties(_this$props2, ["dropdown", "className", "totalCols", "cols", "error", "children", "style"]);

      var labelElem = this.renderLabel();
      var controlElem = this.renderControl({
        children: children,
        dropdown: dropdown,
        error: error
      });
      var formElemChildren = [labelElem, controlElem];
      return this.renderFormElement(_objectSpread(_objectSpread({}, props), {}, {
        className: className,
        error: error,
        totalCols: totalCols,
        cols: cols,
        style: style,
        children: formElemChildren
      }));
    }
  }]);

  return FormElement;
}(React.Component);

_defineProperty(FormElement, "isFormElement", true);
