// react-lightning-design-system/module/scripts/Modal.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classnames from 'classnames';
import { Button } from './Button';
export var ModalHeader = /*#__PURE__*/function (_Component) {
  _inherits(ModalHeader, _Component);

  var _super = _createSuper(ModalHeader);

  function ModalHeader(props) {
    var _this;

    _classCallCheck(this, ModalHeader);

    _this = _super.call(this, props);
    _this.onClose = _this.onClose.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(ModalHeader, [{
    key: "onClose",
    value: function onClose() {
      if (this.props.onClose) {
        this.props.onClose();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          title = _this$props.title,
          tagline = _this$props.tagline,
          closeButton = _this$props.closeButton,
          props = _objectWithoutProperties(_this$props, ["className", "title", "tagline", "closeButton"]);

      delete props.onClose;
      var hdClassNames = classnames(className, 'slds-modal__header');
      return /*#__PURE__*/React.createElement("div", _extends({
        className: hdClassNames
      }, props), /*#__PURE__*/React.createElement("h2", {
        className: "slds-text-heading_medium"
      }, title), tagline ? /*#__PURE__*/React.createElement("p", {
        className: "slds-m-top_x-small"
      }, tagline) : null, closeButton ? /*#__PURE__*/React.createElement(Button, {
        type: "icon-inverse",
        className: "slds-modal__close",
        icon: "close",
        iconSize: "large",
        alt: "Close",
        inverse: true,
        onClick: this.onClose
      }) : null);
    }
  }]);

  return ModalHeader;
}(Component);
export var ModalContent = function ModalContent(_ref) {
  var className = _ref.className,
      children = _ref.children,
      props = _objectWithoutProperties(_ref, ["className", "children"]);

  var ctClassNames = classnames(className, 'slds-modal__content');
  return /*#__PURE__*/React.createElement("div", _extends({
    className: ctClassNames
  }, props), children);
};
export var ModalFooter = function ModalFooter(_ref2) {
  var className = _ref2.className,
      directional = _ref2.directional,
      children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["className", "directional", "children"]);

  var ftClassNames = classnames(className, 'slds-modal__footer', {
    'slds-modal__footer_directional': directional
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    className: ftClassNames
  }, props), children);
};
export var Modal = /*#__PURE__*/function (_Component2) {
  _inherits(Modal, _Component2);

  var _super2 = _createSuper(Modal);

  function Modal(props) {
    var _this2;

    _classCallCheck(this, Modal);

    _this2 = _super2.call(this, props);
    _this2.renderChildComponent = _this2.renderChildComponent.bind(_assertThisInitialized(_this2));
    return _this2;
  }

  _createClass(Modal, [{
    key: "hide",
    value: function hide() {
      if (this.props.onHide) {
        this.props.onHide();
      }
    }
  }, {
    key: "renderChildComponent",
    value: function renderChildComponent(comp) {
      if (comp.type === ModalHeader) {
        return /*#__PURE__*/React.cloneElement(comp, {
          onClose: this.hide.bind(this)
        });
      }

      return comp;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          className = _this$props2.className,
          opened = _this$props2.opened,
          children = _this$props2.children,
          size = _this$props2.size,
          containerStyle = _this$props2.containerStyle,
          props = _objectWithoutProperties(_this$props2, ["className", "opened", "children", "size", "containerStyle"]);

      delete props.onHide;
      var modalClassNames = classnames(className, 'slds-modal', {
        'slds-fade-in-open': opened,
        'slds-modal_large': size === 'large'
      });
      var backdropClassNames = classnames(className, 'slds-backdrop', {
        'slds-backdrop_open': opened
      });
      return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", _extends({
        className: modalClassNames,
        "aria-hidden": !opened,
        role: "dialog"
      }, props), /*#__PURE__*/React.createElement("div", {
        className: "slds-modal__container",
        style: containerStyle
      }, React.Children.map(children, this.renderChildComponent))), /*#__PURE__*/React.createElement("div", {
        className: backdropClassNames
      }));
    }
  }]);

  return Modal;
}(Component);

_defineProperty(Modal, "Header", ModalHeader);

_defineProperty(Modal, "Content", ModalContent);

_defineProperty(Modal, "Footer", ModalFooter);
