// react-lightning-design-system/module/scripts/Notification.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import classnames from 'classnames';
import { Button } from './Button';
import { Icon } from './Icon';
var NOTIFICATION_TYPES = ['alert', 'toast'];
var NOTIFICATION_LEVELS = ['info', 'success', 'warning', 'error'];
export var Notification = function Notification(props) {
  var className = props.className,
      type = props.type,
      level = props.level,
      alt = props.alt,
      _props$alertTexture = props.alertTexture,
      alertTexture = _props$alertTexture === void 0 ? true : _props$alertTexture,
      icon = props.icon,
      _props$iconSize = props.iconSize,
      iconSize = _props$iconSize === void 0 ? 'small' : _props$iconSize,
      onClose = props.onClose,
      children = props.children,
      pprops = _objectWithoutProperties(props, ["className", "type", "level", "alt", "alertTexture", "icon", "iconSize", "onClose", "children"]);

  var typeClassName = type && NOTIFICATION_TYPES.indexOf(type) >= 0 ? "slds-notify_".concat(type) : null;
  var levelClassName = level && NOTIFICATION_LEVELS.indexOf(level) >= 0 ? "slds-theme_".concat(level) : null;
  var alertClassNames = classnames(className, 'slds-notify', typeClassName, levelClassName, alertTexture ? 'slds-theme_alert-texture' : null);
  var iconEl = icon ? /*#__PURE__*/React.createElement(Icon, {
    className: type === 'toast' ? 'slds-m-right_small' : 'slds-m-right_x-small',
    icon: icon,
    size: iconSize,
    fillColor: "none",
    textColor: level === 'warning' ? 'default' : null
  }) : undefined;
  return /*#__PURE__*/React.createElement("div", _extends({
    className: alertClassNames,
    role: "alert"
  }, pprops), alt ? /*#__PURE__*/React.createElement("span", {
    className: "slds-assistive-text"
  }, alt) : undefined, onClose ? /*#__PURE__*/React.createElement(Button, {
    className: "slds-notify__close",
    type: "icon-inverse",
    icon: "close",
    iconSize: type === 'toast' ? 'large' : 'small',
    alt: "Close",
    onClick: onClose
  }) : undefined, type === 'toast' ? /*#__PURE__*/React.createElement("div", {
    className: "slds-notify__content slds-grid"
  }, iconEl, /*#__PURE__*/React.createElement("div", {
    className: "slds-col slds-align-middle"
  }, /*#__PURE__*/React.createElement("h2", {
    className: "slds-text-heading_small"
  }, children))) : /*#__PURE__*/React.createElement("h2", null, iconEl, children));
};
export var Alert = function Alert(props) {
  return /*#__PURE__*/React.createElement(Notification, _extends({}, props, {
    type: "alert"
  }));
};
export var Toast = function Toast(props) {
  return /*#__PURE__*/React.createElement(Notification, _extends({}, props, {
    type: "toast"
  }));
};
