// react-lightning-design-system/module/scripts/PageHeader.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classNames from 'classnames';
import { MediaObject } from './MediaObject';
import { Text } from './Text';
import { Grid, Row, Col } from './Grid';
import { BreadCrumbs, Crumb } from './BreadCrumbs';
export var PageHeaderDetailBody = function PageHeaderDetailBody(_ref) {
  var children = _ref.children,
      props = _objectWithoutProperties(_ref, ["children"]);

  return typeof children === 'string' ? /*#__PURE__*/React.createElement(Text, _extends({
    category: "body",
    type: "regular",
    truncate: true
  }, props), children) : /*#__PURE__*/React.createElement(React.Fragment, null, children);
};
export var PageHeaderDetailLabel = function PageHeaderDetailLabel(_ref2) {
  var children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["children"]);

  return typeof children === 'string' ? /*#__PURE__*/React.createElement(Text, _extends({
    category: "title",
    truncate: true,
    className: "slds-m-bottom_xx-small"
  }, props), children) : /*#__PURE__*/React.createElement(React.Fragment, null, children);
};
export var PageHeaderDetailItem = function PageHeaderDetailItem(props) {
  var children = props.children,
      label = props.label,
      pprops = _objectWithoutProperties(props, ["children", "label"]);

  var manuallyAssembled = !label;
  return /*#__PURE__*/React.createElement("li", _extends({
    className: "slds-page-header__detail-block"
  }, pprops), !manuallyAssembled ? [/*#__PURE__*/React.createElement(PageHeaderDetailLabel, {
    key: 0
  }, label), /*#__PURE__*/React.createElement(PageHeaderDetailBody, {
    key: 1
  }, children)] : [children]);
};
export var PageHeaderDetail = function PageHeaderDetail(_ref3) {
  var children = _ref3.children,
      props = _objectWithoutProperties(_ref3, ["children"]);

  return /*#__PURE__*/React.createElement(Grid, _extends({
    tag: "ul",
    vertical: false,
    className: "slds-page-header__detail-row"
  }, props), children);
};
export var PageHeaderHeadingTitle = function PageHeaderHeadingTitle(props) {
  var className = props.className,
      children = props.children;
  var titleClassNames = classNames(className, 'slds-page-header__title slds-truncate slds-align-middle');
  return /*#__PURE__*/React.createElement("h1", _extends({}, props, {
    className: titleClassNames
  }), children);
};
export var PageHeaderHeading = /*#__PURE__*/function (_Component) {
  _inherits(PageHeaderHeading, _Component);

  var _super = _createSuper(PageHeaderHeading);

  function PageHeaderHeading() {
    _classCallCheck(this, PageHeaderHeading);

    return _super.apply(this, arguments);
  }

  _createClass(PageHeaderHeading, [{
    key: "renderInfo",
    value: function renderInfo(info) {
      return info ? /*#__PURE__*/React.createElement(Text, {
        category: "body",
        type: "small"
      }, info) : null;
    }
  }, {
    key: "renderWithMedia",
    value: function renderWithMedia(figure) {
      var content = this.renderContent();
      return figure ? /*#__PURE__*/React.createElement(MediaObject, {
        figureLeft: figure
      }, content) : content;
    }
  }, {
    key: "renderContent",
    value: function renderContent() {
      var _this$props = this.props,
          rightActions = _this$props.rightActions,
          info = _this$props.info,
          legend = _this$props.legend,
          title = _this$props.title,
          breadCrumbs = _this$props.breadCrumbs,
          leftActions = _this$props.leftActions;
      var infoPart = info && !breadCrumbs && !legend && !rightActions ? this.renderInfo(info) : null;
      var titlePart = typeof title === 'string' ? /*#__PURE__*/React.createElement(PageHeaderHeadingTitle, {
        className: "slds-m-right_small"
      }, title) : title;
      var breadCrumbsPart = null;

      if (breadCrumbs) {
        breadCrumbsPart = breadCrumbs.length && breadCrumbs[0].type === Crumb ? /*#__PURE__*/React.createElement(BreadCrumbs, null, breadCrumbs) : breadCrumbs;
      }

      return /*#__PURE__*/React.createElement("div", null, breadCrumbsPart, legend ? /*#__PURE__*/React.createElement(Text, {
        category: "title",
        type: "caps",
        className: "slds-line-height_reset"
      }, legend) : null, leftActions ? /*#__PURE__*/React.createElement(Grid, {
        vertical: false
      }, titlePart, /*#__PURE__*/React.createElement(Col, {
        className: "slds-shrink-none"
      }, leftActions)) : titlePart, infoPart);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          rightActions = _this$props2.rightActions,
          info = _this$props2.info,
          breadCrumbs = _this$props2.breadCrumbs,
          figure = _this$props2.figure,
          legend = _this$props2.legend;
      var content = this.renderWithMedia(figure);
      var infoPart = info && (breadCrumbs || legend || rightActions) ? this.renderInfo(info) : null;
      return rightActions ? /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Grid, {
        vertical: false
      }, /*#__PURE__*/React.createElement(Col, {
        className: "slds-has-flexi-truncate"
      }, content), /*#__PURE__*/React.createElement(Col, {
        align: "top",
        noFlex: true
      }, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Row, {
        cols: 1
      }, rightActions)))), infoPart) : /*#__PURE__*/React.createElement("div", null, content, infoPart);
    }
  }]);

  return PageHeaderHeading;
}(Component);
export var PageHeader = function PageHeader(props) {
  return /*#__PURE__*/React.createElement("div", _extends({
    className: "slds-page-header",
    role: "banner"
  }, props), props.children);
};
