// react-lightning-design-system/module/scripts/Pill.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classnames from 'classnames';
import { Icon } from './Icon';
import { Button } from './Button';
export var Pill = /*#__PURE__*/function (_Component) {
  _inherits(Pill, _Component);

  var _super = _createSuper(Pill);

  function Pill() {
    var _this;

    _classCallCheck(this, Pill);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "onPillClick", function (e) {
      if (_this.props.onClick) {
        _this.props.onClick(e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onPillRemove", function (e) {
      e.preventDefault();
      e.stopPropagation();

      if (_this.props.onRemove) {
        _this.props.onRemove();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onKeyDown", function (e) {
      if (e.keyCode === 8 || e.keyCode === 46) {
        // Bacspace / DEL
        _this.onPillRemove(e);
      }
    });

    return _this;
  }

  _createClass(Pill, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          icon = _this$props.icon,
          disabled = _this$props.disabled,
          label = _this$props.label,
          tag = _this$props.tag,
          pillRef = _this$props.pillRef,
          truncate = _this$props.truncate,
          className = _this$props.className;
      var Tag = disabled ? 'span' : tag || 'a';
      var pillClassNames = classnames('slds-pill', {
        'slds-truncate': truncate
      }, className);
      return /*#__PURE__*/React.createElement(Tag, {
        ref: function ref(node) {
          if (pillRef) pillRef(node);
        },
        className: pillClassNames,
        onKeyDown: this.onKeyDown,
        onClick: this.onPillClick
      }, icon && icon.icon ? /*#__PURE__*/React.createElement(Icon, {
        className: "slds-pill__icon",
        category: icon.category,
        icon: icon.icon
      }) : undefined, /*#__PURE__*/React.createElement("span", {
        className: "slds-pill__label"
      }, label), /*#__PURE__*/React.createElement(Button, {
        disabled: disabled,
        className: "slds-pill__remove",
        type: "icon-bare",
        icon: "close",
        alt: "Remove",
        tabIndex: -1,
        onClick: this.onPillRemove
      }));
    }
  }]);

  return Pill;
}(Component);
