// react-lightning-design-system/module/scripts/SalesPath.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React from 'react';
import classnames from 'classnames';
import { Icon } from './Icon';

var PathItem = /*#__PURE__*/function (_React$Component) {
  _inherits(PathItem, _React$Component);

  var _super = _createSuper(PathItem);

  function PathItem() {
    _classCallCheck(this, PathItem);

    return _super.apply(this, arguments);
  }

  _createClass(PathItem, [{
    key: "onItemClick",
    value: function onItemClick(itemKey) {
      if (this.props.onSelect) {
        this.props.onSelect(itemKey);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          eventKey = _this$props.eventKey,
          title = _this$props.title,
          completedTitle = _this$props.completedTitle,
          type = _this$props.type;
      var pathItemClassName = classnames('slds-tabs_path__item', "slds-is-".concat(type), className);
      var tabIndex = type === 'current' ? 0 : -1;
      var completedText = completedTitle || 'Stage Complete';
      return /*#__PURE__*/React.createElement("li", {
        className: pathItemClassName,
        role: "presentation"
      }, /*#__PURE__*/React.createElement("a", {
        className: "slds-tabs_path__link",
        "aria-selected": "false",
        tabIndex: tabIndex,
        role: "tab",
        "aria-live": "assertive",
        onClick: this.onItemClick.bind(this, eventKey)
      }, /*#__PURE__*/React.createElement("span", {
        className: "slds-tabs_path__stage"
      }, /*#__PURE__*/React.createElement(Icon, {
        category: "utility",
        icon: "check",
        size: "x-small"
      }), type === 'complete' ? /*#__PURE__*/React.createElement("span", {
        className: "slds-assistive-text"
      }, completedText) : null), /*#__PURE__*/React.createElement("span", {
        className: "slds-tabs_path__title"
      }, title)));
    }
  }]);

  return PathItem;
}(React.Component);

export var SalesPath = /*#__PURE__*/function (_React$Component2) {
  _inherits(SalesPath, _React$Component2);

  var _super2 = _createSuper(SalesPath);

  function SalesPath(props) {
    var _this;

    _classCallCheck(this, SalesPath);

    _this = _super2.call(this, props);
    _this.state = {};
    _this.onItemClick = _this.onItemClick.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(SalesPath, [{
    key: "onItemClick",
    value: function onItemClick(itemKey) {
      if (this.props.onSelect) {
        this.props.onSelect(itemKey);
      } // Uncontrolled


      this.setState({
        activeKey: itemKey
      });
    }
  }, {
    key: "renderSalesPath",
    value: function renderSalesPath(activeKey, paths) {
      var _this2 = this;

      var typeTracker = -1;
      return React.Children.map(paths, function (path) {
        var _path$props = path.props,
            eventKey = _path$props.eventKey,
            type = _path$props.type,
            props = _objectWithoutProperties(_path$props, ["eventKey", "type"]);

        var isActive = eventKey === activeKey;
        typeTracker = isActive ? 0 : typeTracker >= 0 ? 1 : -1;
        var evaluatedType = type || (isActive ? 'current' : typeTracker === -1 ? 'complete' : 'incomplete');
        return /*#__PURE__*/React.createElement(PathItem, _extends({
          eventKey: eventKey,
          type: evaluatedType,
          onSelect: _this2.onItemClick
        }, props));
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          className = _this$props2.className,
          children = _this$props2.children;
      var activeKey = this.props.activeKey || this.state.activeKey || this.props.defaultActiveKey;
      var salesPathClassNames = classnames(className, 'slds-tabs_path');
      return /*#__PURE__*/React.createElement("div", {
        className: salesPathClassNames,
        role: "application tablist"
      }, /*#__PURE__*/React.createElement("ul", {
        className: "slds-tabs_path__nav",
        role: "presentation"
      }, this.renderSalesPath(activeKey, children)));
    }
  }]);

  return SalesPath;
}(React.Component);

_defineProperty(SalesPath, "PathItem", PathItem);
