// react-lightning-design-system/module/scripts/Table.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classnames from 'classnames';
import { Icon } from './Icon';
export var TableHeader = /*#__PURE__*/function (_Component) {
  _inherits(TableHeader, _Component);

  var _super = _createSuper(TableHeader);

  function TableHeader() {
    _classCallCheck(this, TableHeader);

    return _super.apply(this, arguments);
  }

  _createClass(TableHeader, [{
    key: "renderBaseHeaderRow",
    value: function renderBaseHeaderRow() {
      var _this$props = this.props,
          children = _this$props.children,
          sortable = _this$props.sortable,
          hasActions = _this$props.hasActions,
          _this$props$actionsPo = _this$props.actionsPosition,
          actionsPosition = _this$props$actionsPo === void 0 ? 0 : _this$props$actionsPo;
      var nextChildren = [];
      var props = {
        className: 'slds-text-title_caps'
      };
      React.Children.forEach(children.props.children, function (child, index) {
        var childSortable = child.props.sortable;
        nextChildren.push( /*#__PURE__*/React.cloneElement(child, {
          // eslint-disable-next-line react/no-array-index-key
          key: index,
          sortable: typeof childSortable === 'undefined' ? sortable : childSortable
        }));
      });

      if (hasActions) {
        nextChildren = [].concat(_toConsumableArray(nextChildren.slice(0, actionsPosition)), [/*#__PURE__*/React.createElement(TableHeaderColumn, {
          sortable: false,
          width: 50,
          key: -1,
          className: "slds-cell-shrink"
        })], _toConsumableArray(nextChildren.slice(actionsPosition)));
      }

      return /*#__PURE__*/React.cloneElement(children, props, nextChildren);
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement("thead", null, this.renderBaseHeaderRow());
    }
  }]);

  return TableHeader;
}(Component);
export var TableBody = /*#__PURE__*/function (_Component2) {
  _inherits(TableBody, _Component2);

  var _super2 = _createSuper(TableBody);

  function TableBody() {
    _classCallCheck(this, TableBody);

    return _super2.apply(this, arguments);
  }

  _createClass(TableBody, [{
    key: "renderRows",
    value: function renderRows() {
      return React.Children.map(this.props.children, function (child) {
        var children = [];
        React.Children.forEach(child.props.children, function (innerChild, index) {
          if (! /*#__PURE__*/React.isValidElement(innerChild)) return;
          var _ref = innerChild.props,
              truncate = _ref.truncate;
          var props = {
            key: index
          };
          if (typeof truncate !== 'undefined') props.truncate = truncate;
          children.push( /*#__PURE__*/React.cloneElement(innerChild, props));
        });
        return /*#__PURE__*/React.cloneElement(child, {
          className: 'slds-hint-parent'
        }, children);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement("tbody", null, this.renderRows());
    }
  }]);

  return TableBody;
}(Component);
export var TableRow = function TableRow(_ref2) {
  var selected = _ref2.selected,
      props = _objectWithoutProperties(_ref2, ["selected"]);

  var style = props.style;

  if (selected) {
    style = Object.assign({}, style, {
      backgroundColor: '#F8FCF5',
      borderLeft: '2px solid #7db450'
    });
  }

  return /*#__PURE__*/React.createElement("tr", _extends({}, props, {
    style: style
  }), props.children);
};
export var TableHeaderColumn = function TableHeaderColumn(props) {
  var sortable = props.sortable,
      resizable = props.resizable,
      children = props.children,
      className = props.className,
      width = props.width,
      sortDir = props.sortDir,
      onSort = props.onSort,
      sorted = props.sorted,
      align = props.align,
      pprops = _objectWithoutProperties(props, ["sortable", "resizable", "children", "className", "width", "sortDir", "onSort", "sorted", "align"]);

  var oClassNames = classnames(className, 'slds-text-title_caps slds-truncate', _defineProperty({
    'slds-is-sortable': sortable,
    'slds-is-resizable': resizable,
    'slds-is-sorted': sorted
  }, "slds-text-align_".concat(align), align));
  var style = {
    minWidth: width || 'auto'
  };
  var icon = sortDir === 'DESC' ? 'arrowdown' : 'arrowup';
  return /*#__PURE__*/React.createElement("th", _extends({}, pprops, {
    className: oClassNames,
    scope: "col",
    style: style
  }), sortable ? /*#__PURE__*/React.createElement("a", {
    onClick: function onClick(e) {
      e.preventDefault();

      if (onSort) {
        onSort();
      }
    },
    className: "slds-th__action slds-text-link_reset"
  }, /*#__PURE__*/React.createElement("span", {
    className: "slds-assistive-text"
  }, "Sort "), /*#__PURE__*/React.createElement("span", {
    className: "slds-truncate"
  }, children), /*#__PURE__*/React.createElement(Icon, {
    className: "slds-is-sortable__icon",
    textColor: "default",
    container: true,
    size: "x-small",
    category: "utility",
    icon: icon,
    style: {
      position: 'absolute'
    }
  }), /*#__PURE__*/React.createElement("span", {
    className: "slds-assistive-text",
    "aria-live": "assertive",
    "aria-atomic": "true"
  })) : children);
};
export var TableRowColumn = function TableRowColumn(props) {
  var _props$truncate = props.truncate,
      truncate = _props$truncate === void 0 ? true : _props$truncate,
      className = props.className,
      width = props.width,
      children = props.children,
      pprops = _objectWithoutProperties(props, ["truncate", "className", "width", "children"]);

  var oClassNames = classnames(className, {
    'slds-truncate': truncate
  });
  var style = {};
  if (width !== undefined) style.width = width;
  if (!truncate) style.position = 'static';
  return /*#__PURE__*/React.createElement("td", _extends({
    role: "gridcell",
    style: style,
    className: oClassNames
  }, pprops), children);
};
export var TableRowColumnActions = function TableRowColumnActions(props) {
  return /*#__PURE__*/React.createElement(TableRowColumn, {
    className: "slds-cell-shrink",
    "data-label": "Actions",
    truncate: false,
    width: 50,
    style: {
      position: 'static'
    }
  }, props.children);
};
export var Table = /*#__PURE__*/function (_Component3) {
  _inherits(Table, _Component3);

  var _super3 = _createSuper(Table);

  function Table() {
    _classCallCheck(this, Table);

    return _super3.apply(this, arguments);
  }

  _createClass(Table, [{
    key: "onScroll",
    value: function onScroll() {
      var elements = document.getElementsByClassName('react-slds-dropdown-opened');
      if (elements.length) elements[0].childNodes[0].blur();
    }
  }, {
    key: "renderTableHeader",
    value: function renderTableHeader(base) {
      var sortable = this.props.sortable;
      return /*#__PURE__*/React.cloneElement(base, {
        sortable: sortable
      });
    }
  }, {
    key: "renderTableBody",
    value: function renderTableBody(base) {
      return base;
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props2 = this.props,
          className = _this$props2.className,
          bordered = _this$props2.bordered,
          verticalBorders = _this$props2.verticalBorders,
          noRowHover = _this$props2.noRowHover,
          striped = _this$props2.striped,
          fixedLayout = _this$props2.fixedLayout,
          children = _this$props2.children,
          autoWidth = _this$props2.autoWidth,
          wrapperStyle = _this$props2.wrapperStyle,
          pprops = _objectWithoutProperties(_this$props2, ["className", "bordered", "verticalBorders", "noRowHover", "striped", "fixedLayout", "children", "autoWidth", "wrapperStyle"]);

      delete pprops.sortable;
      var tableClassNames = classnames(className, 'slds-table slds-table_cell-buffer', {
        'slds-table_bordered': bordered,
        'slds-no-row-hover': noRowHover,
        'slds-table_striped': striped,
        'slds-table_fixed-layout': fixedLayout,
        'slds-table_col-bordered': verticalBorders
      });
      var wrapStyle = Object.assign({
        overflowY: 'hidden',
        overflowX: 'auto'
      }, wrapperStyle);
      var style = {};
      if (autoWidth) style.width = 'auto';
      var tBody;
      var tHead;
      React.Children.forEach(children, function (child) {
        if (! /*#__PURE__*/React.isValidElement(child)) return;

        if (child.type === TableHeader) {
          tHead = _this.renderTableHeader(child);
        } else if (child.type === TableBody) {
          tBody = _this.renderTableBody(child);
        }
      });
      return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
        style: wrapStyle,
        onScroll: this.onScroll.bind(this)
      }, /*#__PURE__*/React.createElement("table", _extends({
        className: tableClassNames,
        style: style
      }, pprops), tHead, tBody)));
    }
  }]);

  return Table;
}(Component);
