// react-lightning-design-system/module/scripts/Tabs.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classnames from 'classnames';
import { registerStyle } from './util';
import { DropdownButton } from './DropdownButton';

/**
 *
 */
var TabContent = function TabContent(props) {
  var className = props.className,
      active = props.active,
      children = props.children,
      pprops = _objectWithoutProperties(props, ["className", "active", "children"]);

  var tabClassNames = classnames(className, 'slds-tabs__content', "slds-".concat(active ? 'show' : 'hide'));
  return /*#__PURE__*/React.createElement("div", _extends({
    className: tabClassNames,
    role: "tabpanel"
  }, pprops), children);
};

/**
 *
 */
var TabMenu = function TabMenu(props) {
  var _props$icon = props.icon,
      icon = _props$icon === void 0 ? 'down' : _props$icon,
      children = props.children,
      pprops = _objectWithoutProperties(props, ["icon", "children"]);

  return /*#__PURE__*/React.createElement(DropdownButton, _extends({
    className: "react-slds-tab-menu",
    icon: icon,
    type: "icon-bare",
    iconSize: "small",
    nubbinTop: true
  }, pprops), children);
};

var DefaultTabItemRenderer = function DefaultTabItemRenderer(props) {
  return React.Children.only(props.children);
};

/**
 *
 */
var TabItem = function TabItem(props) {
  var type = props.type,
      title = props.title,
      activeKey = props.activeKey,
      eventKey = props.eventKey,
      activeTabRef = props.activeTabRef,
      menu = props.menu,
      menuIcon = props.menuIcon,
      onTabClick = props.onTabClick,
      onTabKeyDown = props.onTabKeyDown;
  var menuItems = props.menuItems;
  menuItems = menu ? menu.props.children : menuItems;
  var menuProps = menu ? menu.props : {};
  var isActive = eventKey === activeKey;
  var tabItemClassName = classnames({
    'slds-tabs__item': !!menuItems
  }, "slds-tabs_".concat(type, "__item"), {
    'slds-active': isActive
  }, {
    'react-slds-tab-with-menu': menu || menuItems
  });
  var tabLinkClassName = "slds-tabs_".concat(type, "__link");

  var _props$tabItemRendere = props.tabItemRenderer,
      TabItemRenderer = _props$tabItemRendere === void 0 ? DefaultTabItemRenderer : _props$tabItemRendere,
      pprops = _objectWithoutProperties(props, ["tabItemRenderer"]);

  return /*#__PURE__*/React.createElement("li", {
    className: tabItemClassName,
    role: "presentation"
  }, /*#__PURE__*/React.createElement(TabItemRenderer, pprops, /*#__PURE__*/React.createElement("span", {
    className: "react-slds-tab-item-content"
  }, /*#__PURE__*/React.createElement("a", {
    className: tabLinkClassName,
    role: "tab",
    ref: isActive ? activeTabRef : undefined,
    tabIndex: isActive ? 0 : -1,
    "aria-selected": isActive,
    onClick: function onClick() {
      return onTabClick && eventKey != null && onTabClick(eventKey);
    },
    onKeyDown: function onKeyDown(e) {
      return onTabKeyDown && eventKey != null && onTabKeyDown(eventKey, e);
    }
  }, title), menuItems ? /*#__PURE__*/React.createElement(TabMenu, _extends({
    icon: menuIcon
  }, menuProps), menuItems) : undefined)));
};

/**
 *
 */
var TabNav = function TabNav(props) {
  var type = props.type,
      tabs = props.tabs,
      activeKey = props.activeKey,
      activeTabRef = props.activeTabRef,
      onTabClick = props.onTabClick,
      onTabKeyDown = props.onTabKeyDown;
  var tabNavClassName = "slds-tabs_".concat(type, "__nav");
  return /*#__PURE__*/React.createElement("ul", {
    className: tabNavClassName,
    role: "tablist"
  }, React.Children.map(tabs, function (tab) {
    return /*#__PURE__*/React.createElement(TabItem, _extends({}, tab.props, {
      type: type,
      activeKey: activeKey,
      activeTabRef: activeTabRef,
      onTabClick: onTabClick,
      onTabKeyDown: onTabKeyDown
    }));
  }));
};

/**
 *
 */
export var Tab = function Tab(props) {
  var className = props.className,
      eventKey = props.eventKey,
      activeKey = props.activeKey,
      children = props.children;
  return /*#__PURE__*/React.createElement(TabContent, {
    className: className,
    active: eventKey === activeKey
  }, children);
};

/**
 *
 */
export var Tabs = /*#__PURE__*/function (_Component) {
  _inherits(Tabs, _Component);

  var _super = _createSuper(Tabs);

  function Tabs(props) {
    var _this;

    _classCallCheck(this, Tabs);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "activeTab", null);

    _defineProperty(_assertThisInitialized(_this), "onTabClick", function (tabKey) {
      if (_this.props.onSelect) {
        _this.props.onSelect(tabKey);
      } // Uncontrolled


      _this.setState({
        activeKey: tabKey,
        focusTab: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onTabKeyDown", function (tabKey, e) {
      if (e.keyCode === 37 || e.keyCode === 39) {
        // left/right cursor key
        var idx = 0;
        var tabKeys = [];
        React.Children.forEach(_this.props.children, function (tab, i) {
          tabKeys.push(tab.props.eventKey);

          if (tabKey === tab.props.eventKey) {
            idx = i;
          }
        });
        var dir = e.keyCode === 37 ? -1 : 1;
        var activeIdx = (idx + dir + tabKeys.length) % tabKeys.length;
        var activeKey = tabKeys[activeIdx];

        _this.onTabClick(activeKey);

        e.preventDefault();
        e.stopPropagation();
      }
    });

    _this.state = {};
    registerStyle('tab-menu', [['.slds-tabs__item.react-slds-tab-with-menu', '{ position: relative !important; overflow: visible !important; }'], ['.slds-tabs__item.react-slds-tab-with-menu > .react-slds-tab-item-content', '{ overflow: hidden }'], ['.slds-tabs__item.react-slds-tab-with-menu > .react-slds-tab-item-content > a', '{ padding-right: 2rem; }'], ['.react-slds-tab-menu', '{ position: absolute; top: 0; right: 0; }'], ['.react-slds-tab-menu button', '{ height: 2.5rem; line-height: 2rem; width: 2rem; visibility: hidden; justify-content: center }'], ['.slds-tabs__item.slds-active .react-slds-tab-menu button', '.slds-tabs__item:hover .react-slds-tab-menu button', '.slds-tabs__item .react-slds-tab-menu button:focus', '{ visibility: visible }']]);
    return _this;
  }

  _createClass(Tabs, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (this.state.focusTab) {
        var el = this.activeTab;

        if (el) {
          el.focus();
        }
        /* eslint-disable react/no-did-update-set-state */


        this.setState({
          focusTab: false
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children;
      var type = this.props.type === 'scoped' ? 'scoped' : 'default';
      var tabsClassNames = classnames(className, "slds-tabs_".concat(type));
      var activeKey = typeof this.props.activeKey !== 'undefined' ? this.props.activeKey : typeof this.state.activeKey !== 'undefined' ? this.state.activeKey : this.props.defaultActiveKey;
      return /*#__PURE__*/React.createElement("div", {
        className: tabsClassNames
      }, /*#__PURE__*/React.createElement(TabNav, {
        type: type,
        activeKey: activeKey,
        activeTabRef: function activeTabRef(node) {
          _this2.activeTab = node;
        },
        tabs: children,
        onTabClick: this.onTabClick,
        onTabKeyDown: this.onTabKeyDown
      }), React.Children.map(children, function (tab) {
        return /*#__PURE__*/React.cloneElement(tab, {
          activeKey: activeKey
        });
      }));
    }
  }]);

  return Tabs;
}(Component);
