// react-lightning-design-system/module/scripts/Toggle.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import classnames from 'classnames';
import { FormElement } from './FormElement';
export var Toggle = /*#__PURE__*/function (_Component) {
  _inherits(Toggle, _Component);

  var _super = _createSuper(Toggle);

  function Toggle() {
    var _this;

    _classCallCheck(this, Toggle);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "node", null);

    return _this;
  }

  _createClass(Toggle, [{
    key: "renderToggle",
    value: function renderToggle(_ref) {
      var className = _ref.className,
          label = _ref.label,
          props = _objectWithoutProperties(_ref, ["className", "label"]);

      var toggleClassNames = classnames(className, 'slds-checkbox_toggle slds-grid');
      return /*#__PURE__*/React.createElement("label", {
        className: toggleClassNames
      }, /*#__PURE__*/React.createElement("span", {
        className: "slds-form-element__label slds-m-bottom_none"
      }, label), /*#__PURE__*/React.createElement("input", _extends({
        name: "checkbox",
        type: "checkbox",
        "aria-describedby": "toggle-desc"
      }, props)), /*#__PURE__*/React.createElement("span", {
        className: "slds-checkbox_faux_container",
        "aria-live": "assertive"
      }, /*#__PURE__*/React.createElement("span", {
        className: "slds-checkbox_faux"
      }), /*#__PURE__*/React.createElement("span", {
        className: "slds-checkbox_on"
      }, "Enabled"), /*#__PURE__*/React.createElement("span", {
        className: "slds-checkbox_off"
      }, "Disabled")));
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          required = _this$props.required,
          error = _this$props.error,
          totalCols = _this$props.totalCols,
          cols = _this$props.cols,
          props = _objectWithoutProperties(_this$props, ["required", "error", "totalCols", "cols"]);

      var formElemProps = {
        required: required,
        error: error,
        totalCols: totalCols,
        cols: cols
      };
      return /*#__PURE__*/React.createElement(FormElement, _extends({
        formElementRef: function formElementRef(node) {
          return _this2.node = node;
        }
      }, formElemProps), this.renderToggle(props));
    }
  }]);

  return Toggle;
}(Component);
