// react-lightning-design-system/lib/scripts/BreadCrumbs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BreadCrumbs = exports.Crumb = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var Crumb = function Crumb(_ref) {
  var className = _ref.className,
      href = _ref.href,
      children = _ref.children,
      props = (0, _objectWithoutProperties2["default"])(_ref, ["className", "href", "children"]);
  var text = children;
  var cClassName = (0, _classnames["default"])('slds-list__item slds-text-heading_label', className);
  return /*#__PURE__*/_react["default"].createElement("li", (0, _extends2["default"])({}, props, {
    className: cClassName
  }), /*#__PURE__*/_react["default"].createElement("a", {
    href: href
  }, text));
};

exports.Crumb = Crumb;

var BreadCrumbs = function BreadCrumbs(_ref2) {
  var label = _ref2.label,
      className = _ref2.className,
      children = _ref2.children,
      props = (0, _objectWithoutProperties2["default"])(_ref2, ["label", "className", "children"]);
  var oClassName = (0, _classnames["default"])('slds-breadcrumb slds-list_horizontal', className);
  return /*#__PURE__*/_react["default"].createElement("nav", (0, _extends2["default"])({}, props, {
    role: "navigation"
  }), label ? /*#__PURE__*/_react["default"].createElement("p", {
    id: "bread-crumb-label",
    className: "slds-assistive-text"
  }, label) : null, /*#__PURE__*/_react["default"].createElement("ol", {
    className: oClassName,
    "aria-labelledby": "bread-crumb-label"
  }, children));
};

exports.BreadCrumbs = BreadCrumbs;
