// react-lightning-design-system/lib/scripts/DropdownMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropdownMenu = exports.MenuItem = exports.DropdownMenuItem = exports.DropdownMenuHandlerContext = exports.MenuHeader = exports.DropdownMenuHeader = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames3 = _interopRequireDefault(require("classnames"));

var _Icon = require("./Icon");

var _AutoAlign = require("./AutoAlign");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var DropdownMenuHeader = function DropdownMenuHeader(_ref) {
  var divider = _ref.divider,
      className = _ref.className,
      children = _ref.children;
  var menuHeaderClass = (0, _classnames3["default"])('slds-dropdown__header', (0, _defineProperty2["default"])({}, "slds-has-divider_".concat(divider, "-space"), divider), className);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: menuHeaderClass
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "slds-text-heading_label"
  }, children));
};

exports.DropdownMenuHeader = DropdownMenuHeader;
var MenuHeader = DropdownMenuHeader;
exports.MenuHeader = MenuHeader;
var DropdownMenuHandlerContext = /*#__PURE__*/(0, _react.createContext)(null);
exports.DropdownMenuHandlerContext = DropdownMenuHandlerContext;

var DropdownMenuItemInner = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DropdownMenuItemInner, _Component);

  var _super = _createSuper(DropdownMenuItemInner);

  function DropdownMenuItemInner() {
    var _this;

    (0, _classCallCheck2["default"])(this, DropdownMenuItemInner);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onKeyDown", function (e) {
      if (e.keyCode === 13 || e.keyCode === 32) {
        // return or space
        e.preventDefault();
        e.stopPropagation();

        _this.onMenuItemClick(e);
      } else if (e.keyCode === 40 || e.keyCode === 38) {
        e.preventDefault();
        e.stopPropagation();
        var currentEl = e.currentTarget.parentElement;
        var itemEl = currentEl ? e.keyCode === 40 ? currentEl.nextSibling : currentEl.previousSibling : null;

        while (itemEl) {
          var anchorEl = itemEl.querySelector('.react-slds-menuitem[tabIndex]');

          if (anchorEl && !anchorEl.disabled) {
            anchorEl.focus();
            return;
          }

          itemEl = e.keyCode === 40 ? itemEl.nextSibling : itemEl.previousSibling;
        }
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onMenuItemClick", function (e) {
      if (_this.props.onClick) {
        _this.props.onClick(e);
      }

      if (_this.props.eventKey != null) {
        _this.props.onMenuSelect(_this.props.eventKey);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onMenuItemBlur", function (e) {
      if (_this.props.onBlur) {
        _this.props.onBlur(e);
      }

      if (_this.props.onMenuBlur) {
        _this.props.onMenuBlur(e);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onMenuItemFocus", function (e) {
      if (_this.props.onFocus) {
        _this.props.onFocus(e);
      }

      if (_this.props.onMenuFocus) {
        _this.props.onMenuFocus(e);
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(DropdownMenuItemInner, [{
    key: "render",
    value: function render() {
      /* eslint-disable @typescript-eslint/no-unused-vars */
      var _this$props = this.props,
          className = _this$props.className,
          label = _this$props.label,
          icon = _this$props.icon,
          iconRight = _this$props.iconRight,
          selected = _this$props.selected,
          disabled = _this$props.disabled,
          divider = _this$props.divider,
          _this$props$tabIndex = _this$props.tabIndex,
          tabIndex = _this$props$tabIndex === void 0 ? 0 : _this$props$tabIndex,
          eventKey = _this$props.eventKey,
          onClick = _this$props.onClick,
          onBlur = _this$props.onBlur,
          onFocus = _this$props.onFocus,
          onMenuSelect = _this$props.onMenuSelect,
          onMenuBlur = _this$props.onMenuBlur,
          onMenuFocus = _this$props.onMenuFocus,
          children = _this$props.children,
          props = (0, _objectWithoutProperties2["default"])(_this$props, ["className", "label", "icon", "iconRight", "selected", "disabled", "divider", "tabIndex", "eventKey", "onClick", "onBlur", "onFocus", "onMenuSelect", "onMenuBlur", "onMenuFocus", "children"]);
      /* eslint-enable @typescript-eslint/no-unused-vars */

      var menuItemClass = (0, _classnames3["default"])('slds-dropdown__item', (0, _defineProperty2["default"])({
        'slds-is-selected': selected
      }, "slds-has-divider_".concat(divider, "-space"), divider), className);
      return /*#__PURE__*/_react["default"].createElement("li", {
        className: menuItemClass
      }, /*#__PURE__*/_react["default"].createElement("a", (0, _extends2["default"])({
        role: "menuitem"
      }, props, {
        className: "slds-truncate react-slds-menuitem",
        "aria-disabled": disabled,
        tabIndex: disabled ? undefined : tabIndex,
        onClick: disabled ? undefined : this.onMenuItemClick,
        onBlur: disabled ? undefined : this.onMenuItemBlur,
        onFocus: disabled ? undefined : this.onMenuItemFocus,
        onKeyDown: disabled ? undefined : this.onKeyDown
      }), /*#__PURE__*/_react["default"].createElement("p", {
        className: "slds-truncate"
      }, icon ? /*#__PURE__*/_react["default"].createElement(_Icon.Icon, {
        icon: icon,
        size: "x-small",
        align: "left"
      }) : null, label || children), iconRight ? /*#__PURE__*/_react["default"].createElement(_Icon.Icon, {
        icon: iconRight,
        size: "x-small",
        align: "right"
      }) : null));
    }
  }]);
  return DropdownMenuItemInner;
}(_react.Component);

var DropdownMenuItem = function DropdownMenuItem(props) {
  return /*#__PURE__*/_react["default"].createElement(DropdownMenuHandlerContext.Consumer, null, function (handlers) {
    return /*#__PURE__*/_react["default"].createElement(DropdownMenuItemInner, (0, _extends2["default"])({}, props, handlers));
  });
};

exports.DropdownMenuItem = DropdownMenuItem;
var MenuItem = DropdownMenuItem;
exports.MenuItem = MenuItem;

var DropdownMenuInner = /*#__PURE__*/function (_Component2) {
  (0, _inherits2["default"])(DropdownMenuInner, _Component2);

  var _super2 = _createSuper(DropdownMenuInner);

  function DropdownMenuInner() {
    var _this2;

    (0, _classCallCheck2["default"])(this, DropdownMenuInner);

    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    _this2 = _super2.call.apply(_super2, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this2), "node", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this2), "handlers", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this2), "onKeyDown", function (e) {
      if (e.keyCode === 27) {
        // ESC
        if (_this2.props.onMenuClose) {
          _this2.props.onMenuClose();
        }
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this2), "onMenuSelect", function (eventKey) {
      if (_this2.props.onMenuSelect) {
        _this2.props.onMenuSelect(eventKey);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this2), "onMenuFocus", function (e) {
      if (_this2.props.onFocus) {
        _this2.props.onFocus(e);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this2), "onMenuBlur", function (e) {
      if (_this2.props.onBlur) {
        _this2.props.onBlur(e);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this2), "getDropdownContext", function () {
      if (!_this2.handlers) {
        _this2.handlers = {
          onMenuSelect: _this2.onMenuSelect,
          onMenuBlur: _this2.onMenuBlur,
          onMenuFocus: _this2.onMenuFocus
        };
      }

      return _this2.handlers;
    });
    return _this2;
  }

  (0, _createClass2["default"])(DropdownMenuInner, [{
    key: "render",
    value: function render() {
      var _this3 = this;

      /* eslint-disable @typescript-eslint/no-unused-vars */
      var _this$props2 = this.props,
          className = _this$props2.className,
          alignment = _this$props2.alignment,
          size = _this$props2.size,
          header = _this$props2.header,
          nubbinTop = _this$props2.nubbinTop,
          hoverPopup = _this$props2.hoverPopup,
          children = _this$props2.children,
          style = _this$props2.style,
          dropdownMenuRef = _this$props2.dropdownMenuRef,
          onFocus = _this$props2.onFocus,
          onBlur = _this$props2.onBlur,
          onMenuSelect = _this$props2.onMenuSelect,
          onMenuClose = _this$props2.onMenuClose,
          rprops = (0, _objectWithoutProperties2["default"])(_this$props2, ["className", "alignment", "size", "header", "nubbinTop", "hoverPopup", "children", "style", "dropdownMenuRef", "onFocus", "onBlur", "onMenuSelect", "onMenuClose"]);
      /* eslint-enable @typescript-eslint/no-unused-vars */

      var nubbin = nubbinTop ? 'auto' : this.props.nubbin;

      var _alignment = (0, _slicedToArray2["default"])(alignment, 2),
          vertAlign = _alignment[0],
          align = _alignment[1];

      var nubbinPosition = nubbin === 'auto' ? "".concat(vertAlign, "-").concat(align) : nubbin;
      var dropdownClassNames = (0, _classnames3["default"])(className, 'slds-dropdown', vertAlign ? "slds-dropdown_".concat(vertAlign) : undefined, align ? "slds-dropdown_".concat(align) : undefined, size ? "slds-dropdown_".concat(size) : undefined, nubbinPosition ? "slds-nubbin_".concat(nubbinPosition) : undefined, {
        'react-slds-no-hover-popup': !hoverPopup
      });

      var handleDOMRef = function handleDOMRef(node) {
        _this3.node = node;

        if (dropdownMenuRef) {
          dropdownMenuRef(node);
        }
      };

      return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
        className: dropdownClassNames,
        ref: handleDOMRef,
        style: _objectSpread({
          outline: 'none'
        }, style),
        onKeyDown: this.onKeyDown,
        tabIndex: -1,
        onFocus: onFocus,
        onBlur: onBlur
      }, rprops), header ? /*#__PURE__*/_react["default"].createElement(MenuHeader, null, header) : null, /*#__PURE__*/_react["default"].createElement("ul", {
        className: "slds-dropdown__list",
        role: "menu"
      }, /*#__PURE__*/_react["default"].createElement(DropdownMenuHandlerContext.Provider, {
        value: this.getDropdownContext()
      }, children)));
    }
  }]);
  return DropdownMenuInner;
}(_react.Component);

function preventPortalizeOnHoverPopup(Cmp) {
  return function (props) {
    return /*#__PURE__*/_react["default"].createElement(Cmp, (0, _extends2["default"])({
      preventPortalize: !!props.hoverPopup
    }, props));
  };
}

var DropdownMenu = preventPortalizeOnHoverPopup((0, _AutoAlign.autoAlign)({
  triggerSelector: '.slds-dropdown-trigger',
  alignmentStyle: 'menu'
})(DropdownMenuInner));
exports.DropdownMenu = DropdownMenu;
