// react-lightning-design-system/lib/scripts/DropdownButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropdownButton = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = require("./Button");

var _DropdownMenu = require("./DropdownMenu");

var _util = require("./util");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var DropdownButton = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(DropdownButton, _Component);

  var _super = _createSuper(DropdownButton);

  function DropdownButton(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, DropdownButton);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "node", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "trigger", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "dropdown", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onBlur", function () {
      setTimeout(function () {
        if (!_this.isFocusedInComponent()) {
          _this.setState({
            opened: false
          });

          if (_this.props.onBlur) {
            _this.props.onBlur();
          }
        }
      }, 10);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onKeyDown", function (e) {
      if (e.keyCode === 40) {
        // down
        e.preventDefault();
        e.stopPropagation();

        if (!_this.state.opened) {
          _this.setState({
            opened: true
          });

          if (_this.props.onClick) {
            _this.props.onClick(e);
          }

          setTimeout(function () {
            _this.focusToTargetItemEl();
          }, 20);
        } else {
          _this.focusToTargetItemEl();
        }
      } else if (e.keyCode === 27) {
        // ESC
        e.preventDefault();
        e.stopPropagation();

        _this.setState({
          opened: false
        });
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onTriggerClick", function (e) {
      if (!_this.props.hoverPopup) {
        _this.setState(function (prevState) {
          return {
            opened: !prevState.opened
          };
        });
      }

      if (_this.props.onClick) {
        _this.props.onClick(e);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onMenuSelect", function (eventKey) {
      if (!_this.props.hoverPopup) {
        setTimeout(function () {
          var triggerElem = _this.trigger;
          if (triggerElem) triggerElem.focus();

          _this.setState({
            opened: false
          });
        }, 10);
      }

      if (_this.props.onMenuSelect) {
        _this.props.onMenuSelect(eventKey);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onMenuClose", function () {
      if (_this.trigger) {
        _this.trigger.focus();
      }

      _this.setState({
        opened: false
      });
    });
    _this.state = {
      opened: false
    };
    (0, _util.registerStyle)('no-hover-popup', [['.slds-dropdown-trigger:hover .slds-dropdown_menu.react-slds-no-hover-popup', '{ visibility: hidden; opacity: 0; }'], ['.slds-dropdown-trigger.react-slds-dropdown-opened .slds-dropdown_menu', '{ visibility: visible !important; opacity: 1 !important; }']]);
    return _this;
  }

  (0, _createClass2["default"])(DropdownButton, [{
    key: "isFocusedInComponent",
    value: function isFocusedInComponent() {
      var targetEl = document.activeElement;
      return (0, _util.isElInChildren)(this.node, targetEl) || (0, _util.isElInChildren)(this.dropdown, targetEl);
    }
  }, {
    key: "focusToTargetItemEl",
    value: function focusToTargetItemEl() {
      var dropdownEl = this.dropdown;

      if (!dropdownEl) {
        return;
      }

      var firstItemEl = dropdownEl.querySelector('.slds-is-selected > .react-slds-menuitem[tabIndex]') || dropdownEl.querySelector('.react-slds-menuitem[tabIndex]');

      if (firstItemEl) {
        firstItemEl.focus();
      }
    }
  }, {
    key: "renderButton",
    value: function renderButton(_ref) {
      var _this2 = this;

      var grouped = _ref.grouped,
          isFirstInGroup = _ref.isFirstInGroup,
          isLastInGroup = _ref.isLastInGroup,
          props = (0, _objectWithoutProperties2["default"])(_ref, ["grouped", "isFirstInGroup", "isLastInGroup"]);
      var pprops = props;
      delete pprops.onMenuItemClick;

      var button = /*#__PURE__*/_react["default"].createElement(_Button.Button, (0, _extends2["default"])({}, pprops, {
        "aria-haspopup": true,
        buttonRef: function buttonRef(node) {
          return _this2.trigger = node;
        },
        onClick: this.onTriggerClick,
        onKeyDown: this.onKeyDown,
        onBlur: this.onBlur
      }));

      if (grouped) {
        var noneStyle = {
          display: 'none'
        };
        return /*#__PURE__*/_react["default"].createElement("div", {
          className: "slds-button-group"
        }, isFirstInGroup ? null : /*#__PURE__*/_react["default"].createElement("button", {
          type: "button",
          className: "slds-button",
          style: noneStyle
        }), button, isLastInGroup ? null : /*#__PURE__*/_react["default"].createElement("button", {
          type: "button",
          className: "slds-button",
          style: noneStyle
        }));
      }

      return button;
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props = this.props,
          className = _this$props.className,
          menuAlign = _this$props.menuAlign,
          menuSize = _this$props.menuSize,
          nubbinTop = _this$props.nubbinTop,
          hoverPopup = _this$props.hoverPopup,
          menuHeader = _this$props.menuHeader,
          type = _this$props.type,
          label = _this$props.label,
          children = _this$props.children,
          style = _this$props.style,
          menuStyle = _this$props.menuStyle,
          props = (0, _objectWithoutProperties2["default"])(_this$props, ["className", "menuAlign", "menuSize", "nubbinTop", "hoverPopup", "menuHeader", "type", "label", "children", "style", "menuStyle"]);
      var icon = this.props.icon;
      var dropdownClassNames = (0, _classnames["default"])(className, 'slds-dropdown-trigger', {
        'slds-button-space-left': !props.grouped,
        'react-slds-dropdown-opened': this.state.opened
      });
      var iconMore = null;

      if (!label && !icon) {
        icon = 'down';
      }

      if (label || type === 'icon-more') {
        iconMore = 'down';
      }

      var dropdown = /*#__PURE__*/_react["default"].createElement(_DropdownMenu.DropdownMenu, {
        portalClassName: className,
        align: menuAlign,
        header: menuHeader,
        size: menuSize,
        nubbinTop: nubbinTop,
        hoverPopup: hoverPopup,
        dropdownMenuRef: function dropdownMenuRef(node) {
          return _this3.dropdown = node;
        },
        onMenuSelect: this.onMenuSelect,
        onMenuClose: this.onMenuClose,
        onBlur: this.onBlur,
        style: Object.assign({
          transition: 'none'
        }, menuStyle)
      }, children);

      return /*#__PURE__*/_react["default"].createElement("div", {
        className: dropdownClassNames,
        style: style,
        ref: function ref(node) {
          return _this3.node = node;
        }
      }, this.renderButton(_objectSpread({
        type: type,
        label: label,
        icon: icon,
        iconMore: iconMore
      }, props)), hoverPopup || this.state.opened ? dropdown : undefined);
    }
  }]);
  return DropdownButton;
}(_react.Component);

exports.DropdownButton = DropdownButton;
