// react-lightning-design-system/lib/scripts/Modal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Modal = exports.ModalFooter = exports.ModalContent = exports.ModalHeader = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = require("./Button");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ModalHeader = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(ModalHeader, _Component);

  var _super = _createSuper(ModalHeader);

  function ModalHeader(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, ModalHeader);
    _this = _super.call(this, props);
    _this.onClose = _this.onClose.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(ModalHeader, [{
    key: "onClose",
    value: function onClose() {
      if (this.props.onClose) {
        this.props.onClose();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          title = _this$props.title,
          tagline = _this$props.tagline,
          closeButton = _this$props.closeButton,
          props = (0, _objectWithoutProperties2["default"])(_this$props, ["className", "title", "tagline", "closeButton"]);
      delete props.onClose;
      var hdClassNames = (0, _classnames["default"])(className, 'slds-modal__header');
      return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
        className: hdClassNames
      }, props), /*#__PURE__*/_react["default"].createElement("h2", {
        className: "slds-text-heading_medium"
      }, title), tagline ? /*#__PURE__*/_react["default"].createElement("p", {
        className: "slds-m-top_x-small"
      }, tagline) : null, closeButton ? /*#__PURE__*/_react["default"].createElement(_Button.Button, {
        type: "icon-inverse",
        className: "slds-modal__close",
        icon: "close",
        iconSize: "large",
        alt: "Close",
        inverse: true,
        onClick: this.onClose
      }) : null);
    }
  }]);
  return ModalHeader;
}(_react.Component);

exports.ModalHeader = ModalHeader;

var ModalContent = function ModalContent(_ref) {
  var className = _ref.className,
      children = _ref.children,
      props = (0, _objectWithoutProperties2["default"])(_ref, ["className", "children"]);
  var ctClassNames = (0, _classnames["default"])(className, 'slds-modal__content');
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: ctClassNames
  }, props), children);
};

exports.ModalContent = ModalContent;

var ModalFooter = function ModalFooter(_ref2) {
  var className = _ref2.className,
      directional = _ref2.directional,
      children = _ref2.children,
      props = (0, _objectWithoutProperties2["default"])(_ref2, ["className", "directional", "children"]);
  var ftClassNames = (0, _classnames["default"])(className, 'slds-modal__footer', {
    'slds-modal__footer_directional': directional
  });
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: ftClassNames
  }, props), children);
};

exports.ModalFooter = ModalFooter;

var Modal = /*#__PURE__*/function (_Component2) {
  (0, _inherits2["default"])(Modal, _Component2);

  var _super2 = _createSuper(Modal);

  function Modal(props) {
    var _this2;

    (0, _classCallCheck2["default"])(this, Modal);
    _this2 = _super2.call(this, props);
    _this2.renderChildComponent = _this2.renderChildComponent.bind((0, _assertThisInitialized2["default"])(_this2));
    return _this2;
  }

  (0, _createClass2["default"])(Modal, [{
    key: "hide",
    value: function hide() {
      if (this.props.onHide) {
        this.props.onHide();
      }
    }
  }, {
    key: "renderChildComponent",
    value: function renderChildComponent(comp) {
      if (comp.type === ModalHeader) {
        return /*#__PURE__*/_react["default"].cloneElement(comp, {
          onClose: this.hide.bind(this)
        });
      }

      return comp;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          className = _this$props2.className,
          opened = _this$props2.opened,
          children = _this$props2.children,
          size = _this$props2.size,
          containerStyle = _this$props2.containerStyle,
          props = (0, _objectWithoutProperties2["default"])(_this$props2, ["className", "opened", "children", "size", "containerStyle"]);
      delete props.onHide;
      var modalClassNames = (0, _classnames["default"])(className, 'slds-modal', {
        'slds-fade-in-open': opened,
        'slds-modal_large': size === 'large'
      });
      var backdropClassNames = (0, _classnames["default"])(className, 'slds-backdrop', {
        'slds-backdrop_open': opened
      });
      return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
        className: modalClassNames,
        "aria-hidden": !opened,
        role: "dialog"
      }, props), /*#__PURE__*/_react["default"].createElement("div", {
        className: "slds-modal__container",
        style: containerStyle
      }, _react["default"].Children.map(children, this.renderChildComponent))), /*#__PURE__*/_react["default"].createElement("div", {
        className: backdropClassNames
      }));
    }
  }]);
  return Modal;
}(_react.Component);

exports.Modal = Modal;
(0, _defineProperty2["default"])(Modal, "Header", ModalHeader);
(0, _defineProperty2["default"])(Modal, "Content", ModalContent);
(0, _defineProperty2["default"])(Modal, "Footer", ModalFooter);
