// react-lightning-design-system/lib/scripts/Notification.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Toast = exports.Alert = exports.Notification = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = require("./Button");

var _Icon = require("./Icon");

var NOTIFICATION_TYPES = ['alert', 'toast'];
var NOTIFICATION_LEVELS = ['info', 'success', 'warning', 'error'];

var Notification = function Notification(props) {
  var className = props.className,
      type = props.type,
      level = props.level,
      alt = props.alt,
      _props$alertTexture = props.alertTexture,
      alertTexture = _props$alertTexture === void 0 ? true : _props$alertTexture,
      icon = props.icon,
      _props$iconSize = props.iconSize,
      iconSize = _props$iconSize === void 0 ? 'small' : _props$iconSize,
      onClose = props.onClose,
      children = props.children,
      pprops = (0, _objectWithoutProperties2["default"])(props, ["className", "type", "level", "alt", "alertTexture", "icon", "iconSize", "onClose", "children"]);
  var typeClassName = type && NOTIFICATION_TYPES.indexOf(type) >= 0 ? "slds-notify_".concat(type) : null;
  var levelClassName = level && NOTIFICATION_LEVELS.indexOf(level) >= 0 ? "slds-theme_".concat(level) : null;
  var alertClassNames = (0, _classnames["default"])(className, 'slds-notify', typeClassName, levelClassName, alertTexture ? 'slds-theme_alert-texture' : null);
  var iconEl = icon ? /*#__PURE__*/_react["default"].createElement(_Icon.Icon, {
    className: type === 'toast' ? 'slds-m-right_small' : 'slds-m-right_x-small',
    icon: icon,
    size: iconSize,
    fillColor: "none",
    textColor: level === 'warning' ? 'default' : null
  }) : undefined;
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: alertClassNames,
    role: "alert"
  }, pprops), alt ? /*#__PURE__*/_react["default"].createElement("span", {
    className: "slds-assistive-text"
  }, alt) : undefined, onClose ? /*#__PURE__*/_react["default"].createElement(_Button.Button, {
    className: "slds-notify__close",
    type: "icon-inverse",
    icon: "close",
    iconSize: type === 'toast' ? 'large' : 'small',
    alt: "Close",
    onClick: onClose
  }) : undefined, type === 'toast' ? /*#__PURE__*/_react["default"].createElement("div", {
    className: "slds-notify__content slds-grid"
  }, iconEl, /*#__PURE__*/_react["default"].createElement("div", {
    className: "slds-col slds-align-middle"
  }, /*#__PURE__*/_react["default"].createElement("h2", {
    className: "slds-text-heading_small"
  }, children))) : /*#__PURE__*/_react["default"].createElement("h2", null, iconEl, children));
};

exports.Notification = Notification;

var Alert = function Alert(props) {
  return /*#__PURE__*/_react["default"].createElement(Notification, (0, _extends2["default"])({}, props, {
    type: "alert"
  }));
};

exports.Alert = Alert;

var Toast = function Toast(props) {
  return /*#__PURE__*/_react["default"].createElement(Notification, (0, _extends2["default"])({}, props, {
    type: "toast"
  }));
};

exports.Toast = Toast;
