// react-lightning-design-system/lib/scripts/PageHeader.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PageHeader = exports.PageHeaderHeading = exports.PageHeaderHeadingTitle = exports.PageHeaderDetail = exports.PageHeaderDetailItem = exports.PageHeaderDetailLabel = exports.PageHeaderDetailBody = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _MediaObject = require("./MediaObject");

var _Text = require("./Text");

var _Grid = require("./Grid");

var _BreadCrumbs = require("./BreadCrumbs");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var PageHeaderDetailBody = function PageHeaderDetailBody(_ref) {
  var children = _ref.children,
      props = (0, _objectWithoutProperties2["default"])(_ref, ["children"]);
  return typeof children === 'string' ? /*#__PURE__*/_react["default"].createElement(_Text.Text, (0, _extends2["default"])({
    category: "body",
    type: "regular",
    truncate: true
  }, props), children) : /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, children);
};

exports.PageHeaderDetailBody = PageHeaderDetailBody;

var PageHeaderDetailLabel = function PageHeaderDetailLabel(_ref2) {
  var children = _ref2.children,
      props = (0, _objectWithoutProperties2["default"])(_ref2, ["children"]);
  return typeof children === 'string' ? /*#__PURE__*/_react["default"].createElement(_Text.Text, (0, _extends2["default"])({
    category: "title",
    truncate: true,
    className: "slds-m-bottom_xx-small"
  }, props), children) : /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, children);
};

exports.PageHeaderDetailLabel = PageHeaderDetailLabel;

var PageHeaderDetailItem = function PageHeaderDetailItem(props) {
  var children = props.children,
      label = props.label,
      pprops = (0, _objectWithoutProperties2["default"])(props, ["children", "label"]);
  var manuallyAssembled = !label;
  return /*#__PURE__*/_react["default"].createElement("li", (0, _extends2["default"])({
    className: "slds-page-header__detail-block"
  }, pprops), !manuallyAssembled ? [/*#__PURE__*/_react["default"].createElement(PageHeaderDetailLabel, {
    key: 0
  }, label), /*#__PURE__*/_react["default"].createElement(PageHeaderDetailBody, {
    key: 1
  }, children)] : [children]);
};

exports.PageHeaderDetailItem = PageHeaderDetailItem;

var PageHeaderDetail = function PageHeaderDetail(_ref3) {
  var children = _ref3.children,
      props = (0, _objectWithoutProperties2["default"])(_ref3, ["children"]);
  return /*#__PURE__*/_react["default"].createElement(_Grid.Grid, (0, _extends2["default"])({
    tag: "ul",
    vertical: false,
    className: "slds-page-header__detail-row"
  }, props), children);
};

exports.PageHeaderDetail = PageHeaderDetail;

var PageHeaderHeadingTitle = function PageHeaderHeadingTitle(props) {
  var className = props.className,
      children = props.children;
  var titleClassNames = (0, _classnames["default"])(className, 'slds-page-header__title slds-truncate slds-align-middle');
  return /*#__PURE__*/_react["default"].createElement("h1", (0, _extends2["default"])({}, props, {
    className: titleClassNames
  }), children);
};

exports.PageHeaderHeadingTitle = PageHeaderHeadingTitle;

var PageHeaderHeading = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(PageHeaderHeading, _Component);

  var _super = _createSuper(PageHeaderHeading);

  function PageHeaderHeading() {
    (0, _classCallCheck2["default"])(this, PageHeaderHeading);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(PageHeaderHeading, [{
    key: "renderInfo",
    value: function renderInfo(info) {
      return info ? /*#__PURE__*/_react["default"].createElement(_Text.Text, {
        category: "body",
        type: "small"
      }, info) : null;
    }
  }, {
    key: "renderWithMedia",
    value: function renderWithMedia(figure) {
      var content = this.renderContent();
      return figure ? /*#__PURE__*/_react["default"].createElement(_MediaObject.MediaObject, {
        figureLeft: figure
      }, content) : content;
    }
  }, {
    key: "renderContent",
    value: function renderContent() {
      var _this$props = this.props,
          rightActions = _this$props.rightActions,
          info = _this$props.info,
          legend = _this$props.legend,
          title = _this$props.title,
          breadCrumbs = _this$props.breadCrumbs,
          leftActions = _this$props.leftActions;
      var infoPart = info && !breadCrumbs && !legend && !rightActions ? this.renderInfo(info) : null;
      var titlePart = typeof title === 'string' ? /*#__PURE__*/_react["default"].createElement(PageHeaderHeadingTitle, {
        className: "slds-m-right_small"
      }, title) : title;
      var breadCrumbsPart = null;

      if (breadCrumbs) {
        breadCrumbsPart = breadCrumbs.length && breadCrumbs[0].type === _BreadCrumbs.Crumb ? /*#__PURE__*/_react["default"].createElement(_BreadCrumbs.BreadCrumbs, null, breadCrumbs) : breadCrumbs;
      }

      return /*#__PURE__*/_react["default"].createElement("div", null, breadCrumbsPart, legend ? /*#__PURE__*/_react["default"].createElement(_Text.Text, {
        category: "title",
        type: "caps",
        className: "slds-line-height_reset"
      }, legend) : null, leftActions ? /*#__PURE__*/_react["default"].createElement(_Grid.Grid, {
        vertical: false
      }, titlePart, /*#__PURE__*/_react["default"].createElement(_Grid.Col, {
        className: "slds-shrink-none"
      }, leftActions)) : titlePart, infoPart);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          rightActions = _this$props2.rightActions,
          info = _this$props2.info,
          breadCrumbs = _this$props2.breadCrumbs,
          figure = _this$props2.figure,
          legend = _this$props2.legend;
      var content = this.renderWithMedia(figure);
      var infoPart = info && (breadCrumbs || legend || rightActions) ? this.renderInfo(info) : null;
      return rightActions ? /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement(_Grid.Grid, {
        vertical: false
      }, /*#__PURE__*/_react["default"].createElement(_Grid.Col, {
        className: "slds-has-flexi-truncate"
      }, content), /*#__PURE__*/_react["default"].createElement(_Grid.Col, {
        align: "top",
        noFlex: true
      }, /*#__PURE__*/_react["default"].createElement(_Grid.Grid, null, /*#__PURE__*/_react["default"].createElement(_Grid.Row, {
        cols: 1
      }, rightActions)))), infoPart) : /*#__PURE__*/_react["default"].createElement("div", null, content, infoPart);
    }
  }]);
  return PageHeaderHeading;
}(_react.Component);

exports.PageHeaderHeading = PageHeaderHeading;

var PageHeader = function PageHeader(props) {
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: "slds-page-header",
    role: "banner"
  }, props), props.children);
};

exports.PageHeader = PageHeader;
