// react-lightning-design-system/lib/scripts/Pill.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pill = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icon = require("./Icon");

var _Button = require("./Button");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Pill = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Pill, _Component);

  var _super = _createSuper(Pill);

  function Pill() {
    var _this;

    (0, _classCallCheck2["default"])(this, Pill);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onPillClick", function (e) {
      if (_this.props.onClick) {
        _this.props.onClick(e);
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onPillRemove", function (e) {
      e.preventDefault();
      e.stopPropagation();

      if (_this.props.onRemove) {
        _this.props.onRemove();
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onKeyDown", function (e) {
      if (e.keyCode === 8 || e.keyCode === 46) {
        // Bacspace / DEL
        _this.onPillRemove(e);
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(Pill, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          icon = _this$props.icon,
          disabled = _this$props.disabled,
          label = _this$props.label,
          tag = _this$props.tag,
          pillRef = _this$props.pillRef,
          truncate = _this$props.truncate,
          className = _this$props.className;
      var Tag = disabled ? 'span' : tag || 'a';
      var pillClassNames = (0, _classnames["default"])('slds-pill', {
        'slds-truncate': truncate
      }, className);
      return /*#__PURE__*/_react["default"].createElement(Tag, {
        ref: function ref(node) {
          if (pillRef) pillRef(node);
        },
        className: pillClassNames,
        onKeyDown: this.onKeyDown,
        onClick: this.onPillClick
      }, icon && icon.icon ? /*#__PURE__*/_react["default"].createElement(_Icon.Icon, {
        className: "slds-pill__icon",
        category: icon.category,
        icon: icon.icon
      }) : undefined, /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-pill__label"
      }, label), /*#__PURE__*/_react["default"].createElement(_Button.Button, {
        disabled: disabled,
        className: "slds-pill__remove",
        type: "icon-bare",
        icon: "close",
        alt: "Remove",
        tabIndex: -1,
        onClick: this.onPillRemove
      }));
    }
  }]);
  return Pill;
}(_react.Component);

exports.Pill = Pill;
