// react-lightning-design-system/lib/scripts/Popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Popover = exports.PopoverBody = exports.PopoverHeader = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _AutoAlign = require("./AutoAlign");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var PopoverHeader = function PopoverHeader(props) {
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "slds-popover__header"
  }, props.children);
};

exports.PopoverHeader = PopoverHeader;

var PopoverBody = function PopoverBody(props) {
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: "slds-popover__body"
  }, props), props.children);
};

exports.PopoverBody = PopoverBody;

var PopoverInner = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(PopoverInner, _React$Component);

  var _super = _createSuper(PopoverInner);

  function PopoverInner() {
    var _this;

    (0, _classCallCheck2["default"])(this, PopoverInner);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "node", null);
    return _this;
  }

  (0, _createClass2["default"])(PopoverInner, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          children = _this$props.children,
          alignment = _this$props.alignment,
          hidden = _this$props.hidden,
          theme = _this$props.theme,
          tooltip = _this$props.tooltip,
          style = _this$props.style,
          bodyStyle = _this$props.bodyStyle,
          props = (0, _objectWithoutProperties2["default"])(_this$props, ["children", "alignment", "hidden", "theme", "tooltip", "style", "bodyStyle"]);
      var nubbinPosition = alignment.join('-');

      var _alignment = (0, _slicedToArray2["default"])(alignment, 2),
          firstAlign = _alignment[0],
          secondAlign = _alignment[1];

      var popoverClassNames = (0, _classnames["default"])('slds-popover', {
        'slds-hide': hidden,
        'slds-popover_tooltip': tooltip
      }, "slds-nubbin_".concat(nubbinPosition), "slds-m-".concat(firstAlign, "_small"), theme ? "slds-theme_".concat(theme) : undefined);

      var rootStyle = _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, style), {}, (0, _defineProperty2["default"])({
        position: 'absolute'
      }, firstAlign, 0), secondAlign ? (0, _defineProperty2["default"])({}, secondAlign, 0) : {}), tooltip ? {
        width: 'max-content'
      } : {}), {}, {
        transform: secondAlign === undefined ? firstAlign === 'top' || firstAlign === 'bottom' ? 'translateX(-50%)' : firstAlign === 'left' || firstAlign === 'right' ? 'translateY(-50%)' : undefined : undefined
      });

      return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
        ref: function ref(node) {
          return _this2.node = node;
        },
        className: popoverClassNames,
        role: tooltip ? 'tooltip' : 'dialog',
        style: rootStyle
      }, props), /*#__PURE__*/_react["default"].createElement(PopoverBody, {
        style: bodyStyle
      }, children));
    }
  }]);
  return PopoverInner;
}(_react["default"].Component);
/**
 *
 */


function map(Cmp, fn) {
  return function (p1) {
    return /*#__PURE__*/_react["default"].createElement(Cmp, fn(p1));
  };
}
/**
 *
 */


var Popover = map((0, _AutoAlign.autoAlign)({
  triggerSelector: '.slds-dropdown-trigger',
  alignmentStyle: 'popover'
})(PopoverInner), function (_ref2) {
  var position = _ref2.position,
      props = (0, _objectWithoutProperties2["default"])(_ref2, ["position"]);
  var alignment = position ? position.split('-') : undefined;
  return _objectSpread({
    alignment: alignment
  }, props);
});
exports.Popover = Popover;
Popover.displayName = 'Popover';
