// react-lightning-design-system/lib/scripts/SalesPath.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SalesPath = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icon = require("./Icon");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var PathItem = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(PathItem, _React$Component);

  var _super = _createSuper(PathItem);

  function PathItem() {
    (0, _classCallCheck2["default"])(this, PathItem);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(PathItem, [{
    key: "onItemClick",
    value: function onItemClick(itemKey) {
      if (this.props.onSelect) {
        this.props.onSelect(itemKey);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          eventKey = _this$props.eventKey,
          title = _this$props.title,
          completedTitle = _this$props.completedTitle,
          type = _this$props.type;
      var pathItemClassName = (0, _classnames["default"])('slds-tabs_path__item', "slds-is-".concat(type), className);
      var tabIndex = type === 'current' ? 0 : -1;
      var completedText = completedTitle || 'Stage Complete';
      return /*#__PURE__*/_react["default"].createElement("li", {
        className: pathItemClassName,
        role: "presentation"
      }, /*#__PURE__*/_react["default"].createElement("a", {
        className: "slds-tabs_path__link",
        "aria-selected": "false",
        tabIndex: tabIndex,
        role: "tab",
        "aria-live": "assertive",
        onClick: this.onItemClick.bind(this, eventKey)
      }, /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-tabs_path__stage"
      }, /*#__PURE__*/_react["default"].createElement(_Icon.Icon, {
        category: "utility",
        icon: "check",
        size: "x-small"
      }), type === 'complete' ? /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-assistive-text"
      }, completedText) : null), /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-tabs_path__title"
      }, title)));
    }
  }]);
  return PathItem;
}(_react["default"].Component);

var SalesPath = /*#__PURE__*/function (_React$Component2) {
  (0, _inherits2["default"])(SalesPath, _React$Component2);

  var _super2 = _createSuper(SalesPath);

  function SalesPath(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, SalesPath);
    _this = _super2.call(this, props);
    _this.state = {};
    _this.onItemClick = _this.onItemClick.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(SalesPath, [{
    key: "onItemClick",
    value: function onItemClick(itemKey) {
      if (this.props.onSelect) {
        this.props.onSelect(itemKey);
      } // Uncontrolled


      this.setState({
        activeKey: itemKey
      });
    }
  }, {
    key: "renderSalesPath",
    value: function renderSalesPath(activeKey, paths) {
      var _this2 = this;

      var typeTracker = -1;
      return _react["default"].Children.map(paths, function (path) {
        var _path$props = path.props,
            eventKey = _path$props.eventKey,
            type = _path$props.type,
            props = (0, _objectWithoutProperties2["default"])(_path$props, ["eventKey", "type"]);
        var isActive = eventKey === activeKey;
        typeTracker = isActive ? 0 : typeTracker >= 0 ? 1 : -1;
        var evaluatedType = type || (isActive ? 'current' : typeTracker === -1 ? 'complete' : 'incomplete');
        return /*#__PURE__*/_react["default"].createElement(PathItem, (0, _extends2["default"])({
          eventKey: eventKey,
          type: evaluatedType,
          onSelect: _this2.onItemClick
        }, props));
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          className = _this$props2.className,
          children = _this$props2.children;
      var activeKey = this.props.activeKey || this.state.activeKey || this.props.defaultActiveKey;
      var salesPathClassNames = (0, _classnames["default"])(className, 'slds-tabs_path');
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: salesPathClassNames,
        role: "application tablist"
      }, /*#__PURE__*/_react["default"].createElement("ul", {
        className: "slds-tabs_path__nav",
        role: "presentation"
      }, this.renderSalesPath(activeKey, children)));
    }
  }]);
  return SalesPath;
}(_react["default"].Component);

exports.SalesPath = SalesPath;
(0, _defineProperty2["default"])(SalesPath, "PathItem", PathItem);
