// react-lightning-design-system/lib/scripts/Table.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Table = exports.TableRowColumnActions = exports.TableRowColumn = exports.TableHeaderColumn = exports.TableRow = exports.TableBody = exports.TableHeader = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _Icon = require("./Icon");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var TableHeader = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(TableHeader, _Component);

  var _super = _createSuper(TableHeader);

  function TableHeader() {
    (0, _classCallCheck2["default"])(this, TableHeader);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(TableHeader, [{
    key: "renderBaseHeaderRow",
    value: function renderBaseHeaderRow() {
      var _this$props = this.props,
          children = _this$props.children,
          sortable = _this$props.sortable,
          hasActions = _this$props.hasActions,
          _this$props$actionsPo = _this$props.actionsPosition,
          actionsPosition = _this$props$actionsPo === void 0 ? 0 : _this$props$actionsPo;
      var nextChildren = [];
      var props = {
        className: 'slds-text-title_caps'
      };

      _react["default"].Children.forEach(children.props.children, function (child, index) {
        var childSortable = child.props.sortable;
        nextChildren.push( /*#__PURE__*/_react["default"].cloneElement(child, {
          // eslint-disable-next-line react/no-array-index-key
          key: index,
          sortable: typeof childSortable === 'undefined' ? sortable : childSortable
        }));
      });

      if (hasActions) {
        nextChildren = [].concat((0, _toConsumableArray2["default"])(nextChildren.slice(0, actionsPosition)), [/*#__PURE__*/_react["default"].createElement(TableHeaderColumn, {
          sortable: false,
          width: 50,
          key: -1,
          className: "slds-cell-shrink"
        })], (0, _toConsumableArray2["default"])(nextChildren.slice(actionsPosition)));
      }

      return /*#__PURE__*/_react["default"].cloneElement(children, props, nextChildren);
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react["default"].createElement("thead", null, this.renderBaseHeaderRow());
    }
  }]);
  return TableHeader;
}(_react.Component);

exports.TableHeader = TableHeader;

var TableBody = /*#__PURE__*/function (_Component2) {
  (0, _inherits2["default"])(TableBody, _Component2);

  var _super2 = _createSuper(TableBody);

  function TableBody() {
    (0, _classCallCheck2["default"])(this, TableBody);
    return _super2.apply(this, arguments);
  }

  (0, _createClass2["default"])(TableBody, [{
    key: "renderRows",
    value: function renderRows() {
      return _react["default"].Children.map(this.props.children, function (child) {
        var children = [];

        _react["default"].Children.forEach(child.props.children, function (innerChild, index) {
          if (! /*#__PURE__*/_react["default"].isValidElement(innerChild)) return;
          var _ref = innerChild.props,
              truncate = _ref.truncate;
          var props = {
            key: index
          };
          if (typeof truncate !== 'undefined') props.truncate = truncate;
          children.push( /*#__PURE__*/_react["default"].cloneElement(innerChild, props));
        });

        return /*#__PURE__*/_react["default"].cloneElement(child, {
          className: 'slds-hint-parent'
        }, children);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react["default"].createElement("tbody", null, this.renderRows());
    }
  }]);
  return TableBody;
}(_react.Component);

exports.TableBody = TableBody;

var TableRow = function TableRow(_ref2) {
  var selected = _ref2.selected,
      props = (0, _objectWithoutProperties2["default"])(_ref2, ["selected"]);
  var style = props.style;

  if (selected) {
    style = Object.assign({}, style, {
      backgroundColor: '#F8FCF5',
      borderLeft: '2px solid #7db450'
    });
  }

  return /*#__PURE__*/_react["default"].createElement("tr", (0, _extends2["default"])({}, props, {
    style: style
  }), props.children);
};

exports.TableRow = TableRow;

var TableHeaderColumn = function TableHeaderColumn(props) {
  var sortable = props.sortable,
      resizable = props.resizable,
      children = props.children,
      className = props.className,
      width = props.width,
      sortDir = props.sortDir,
      onSort = props.onSort,
      sorted = props.sorted,
      align = props.align,
      pprops = (0, _objectWithoutProperties2["default"])(props, ["sortable", "resizable", "children", "className", "width", "sortDir", "onSort", "sorted", "align"]);
  var oClassNames = (0, _classnames2["default"])(className, 'slds-text-title_caps slds-truncate', (0, _defineProperty2["default"])({
    'slds-is-sortable': sortable,
    'slds-is-resizable': resizable,
    'slds-is-sorted': sorted
  }, "slds-text-align_".concat(align), align));
  var style = {
    minWidth: width || 'auto'
  };
  var icon = sortDir === 'DESC' ? 'arrowdown' : 'arrowup';
  return /*#__PURE__*/_react["default"].createElement("th", (0, _extends2["default"])({}, pprops, {
    className: oClassNames,
    scope: "col",
    style: style
  }), sortable ? /*#__PURE__*/_react["default"].createElement("a", {
    onClick: function onClick(e) {
      e.preventDefault();

      if (onSort) {
        onSort();
      }
    },
    className: "slds-th__action slds-text-link_reset"
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "slds-assistive-text"
  }, "Sort "), /*#__PURE__*/_react["default"].createElement("span", {
    className: "slds-truncate"
  }, children), /*#__PURE__*/_react["default"].createElement(_Icon.Icon, {
    className: "slds-is-sortable__icon",
    textColor: "default",
    container: true,
    size: "x-small",
    category: "utility",
    icon: icon,
    style: {
      position: 'absolute'
    }
  }), /*#__PURE__*/_react["default"].createElement("span", {
    className: "slds-assistive-text",
    "aria-live": "assertive",
    "aria-atomic": "true"
  })) : children);
};

exports.TableHeaderColumn = TableHeaderColumn;

var TableRowColumn = function TableRowColumn(props) {
  var _props$truncate = props.truncate,
      truncate = _props$truncate === void 0 ? true : _props$truncate,
      className = props.className,
      width = props.width,
      children = props.children,
      pprops = (0, _objectWithoutProperties2["default"])(props, ["truncate", "className", "width", "children"]);
  var oClassNames = (0, _classnames2["default"])(className, {
    'slds-truncate': truncate
  });
  var style = {};
  if (width !== undefined) style.width = width;
  if (!truncate) style.position = 'static';
  return /*#__PURE__*/_react["default"].createElement("td", (0, _extends2["default"])({
    role: "gridcell",
    style: style,
    className: oClassNames
  }, pprops), children);
};

exports.TableRowColumn = TableRowColumn;

var TableRowColumnActions = function TableRowColumnActions(props) {
  return /*#__PURE__*/_react["default"].createElement(TableRowColumn, {
    className: "slds-cell-shrink",
    "data-label": "Actions",
    truncate: false,
    width: 50,
    style: {
      position: 'static'
    }
  }, props.children);
};

exports.TableRowColumnActions = TableRowColumnActions;

var Table = /*#__PURE__*/function (_Component3) {
  (0, _inherits2["default"])(Table, _Component3);

  var _super3 = _createSuper(Table);

  function Table() {
    (0, _classCallCheck2["default"])(this, Table);
    return _super3.apply(this, arguments);
  }

  (0, _createClass2["default"])(Table, [{
    key: "onScroll",
    value: function onScroll() {
      var elements = document.getElementsByClassName('react-slds-dropdown-opened');
      if (elements.length) elements[0].childNodes[0].blur();
    }
  }, {
    key: "renderTableHeader",
    value: function renderTableHeader(base) {
      var sortable = this.props.sortable;
      return /*#__PURE__*/_react["default"].cloneElement(base, {
        sortable: sortable
      });
    }
  }, {
    key: "renderTableBody",
    value: function renderTableBody(base) {
      return base;
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props2 = this.props,
          className = _this$props2.className,
          bordered = _this$props2.bordered,
          verticalBorders = _this$props2.verticalBorders,
          noRowHover = _this$props2.noRowHover,
          striped = _this$props2.striped,
          fixedLayout = _this$props2.fixedLayout,
          children = _this$props2.children,
          autoWidth = _this$props2.autoWidth,
          wrapperStyle = _this$props2.wrapperStyle,
          pprops = (0, _objectWithoutProperties2["default"])(_this$props2, ["className", "bordered", "verticalBorders", "noRowHover", "striped", "fixedLayout", "children", "autoWidth", "wrapperStyle"]);
      delete pprops.sortable;
      var tableClassNames = (0, _classnames2["default"])(className, 'slds-table slds-table_cell-buffer', {
        'slds-table_bordered': bordered,
        'slds-no-row-hover': noRowHover,
        'slds-table_striped': striped,
        'slds-table_fixed-layout': fixedLayout,
        'slds-table_col-bordered': verticalBorders
      });
      var wrapStyle = Object.assign({
        overflowY: 'hidden',
        overflowX: 'auto'
      }, wrapperStyle);
      var style = {};
      if (autoWidth) style.width = 'auto';
      var tBody;
      var tHead;

      _react["default"].Children.forEach(children, function (child) {
        if (! /*#__PURE__*/_react["default"].isValidElement(child)) return;

        if (child.type === TableHeader) {
          tHead = _this.renderTableHeader(child);
        } else if (child.type === TableBody) {
          tBody = _this.renderTableBody(child);
        }
      });

      return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("div", {
        style: wrapStyle,
        onScroll: this.onScroll.bind(this)
      }, /*#__PURE__*/_react["default"].createElement("table", (0, _extends2["default"])({
        className: tableClassNames,
        style: style
      }, pprops), tHead, tBody)));
    }
  }]);
  return Table;
}(_react.Component);

exports.Table = Table;
