// react-lightning-design-system/lib/scripts/Tabs.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tabs = exports.Tab = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _util = require("./util");

var _DropdownButton = require("./DropdownButton");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 *
 */
var TabContent = function TabContent(props) {
  var className = props.className,
      active = props.active,
      children = props.children,
      pprops = (0, _objectWithoutProperties2["default"])(props, ["className", "active", "children"]);
  var tabClassNames = (0, _classnames["default"])(className, 'slds-tabs__content', "slds-".concat(active ? 'show' : 'hide'));
  return /*#__PURE__*/_react["default"].createElement("div", (0, _extends2["default"])({
    className: tabClassNames,
    role: "tabpanel"
  }, pprops), children);
};

/**
 *
 */
var TabMenu = function TabMenu(props) {
  var _props$icon = props.icon,
      icon = _props$icon === void 0 ? 'down' : _props$icon,
      children = props.children,
      pprops = (0, _objectWithoutProperties2["default"])(props, ["icon", "children"]);
  return /*#__PURE__*/_react["default"].createElement(_DropdownButton.DropdownButton, (0, _extends2["default"])({
    className: "react-slds-tab-menu",
    icon: icon,
    type: "icon-bare",
    iconSize: "small",
    nubbinTop: true
  }, pprops), children);
};

var DefaultTabItemRenderer = function DefaultTabItemRenderer(props) {
  return _react["default"].Children.only(props.children);
};

/**
 *
 */
var TabItem = function TabItem(props) {
  var type = props.type,
      title = props.title,
      activeKey = props.activeKey,
      eventKey = props.eventKey,
      activeTabRef = props.activeTabRef,
      menu = props.menu,
      menuIcon = props.menuIcon,
      onTabClick = props.onTabClick,
      onTabKeyDown = props.onTabKeyDown;
  var menuItems = props.menuItems;
  menuItems = menu ? menu.props.children : menuItems;
  var menuProps = menu ? menu.props : {};
  var isActive = eventKey === activeKey;
  var tabItemClassName = (0, _classnames["default"])({
    'slds-tabs__item': !!menuItems
  }, "slds-tabs_".concat(type, "__item"), {
    'slds-active': isActive
  }, {
    'react-slds-tab-with-menu': menu || menuItems
  });
  var tabLinkClassName = "slds-tabs_".concat(type, "__link");
  var _props$tabItemRendere = props.tabItemRenderer,
      TabItemRenderer = _props$tabItemRendere === void 0 ? DefaultTabItemRenderer : _props$tabItemRendere,
      pprops = (0, _objectWithoutProperties2["default"])(props, ["tabItemRenderer"]);
  return /*#__PURE__*/_react["default"].createElement("li", {
    className: tabItemClassName,
    role: "presentation"
  }, /*#__PURE__*/_react["default"].createElement(TabItemRenderer, pprops, /*#__PURE__*/_react["default"].createElement("span", {
    className: "react-slds-tab-item-content"
  }, /*#__PURE__*/_react["default"].createElement("a", {
    className: tabLinkClassName,
    role: "tab",
    ref: isActive ? activeTabRef : undefined,
    tabIndex: isActive ? 0 : -1,
    "aria-selected": isActive,
    onClick: function onClick() {
      return onTabClick && eventKey != null && onTabClick(eventKey);
    },
    onKeyDown: function onKeyDown(e) {
      return onTabKeyDown && eventKey != null && onTabKeyDown(eventKey, e);
    }
  }, title), menuItems ? /*#__PURE__*/_react["default"].createElement(TabMenu, (0, _extends2["default"])({
    icon: menuIcon
  }, menuProps), menuItems) : undefined)));
};

/**
 *
 */
var TabNav = function TabNav(props) {
  var type = props.type,
      tabs = props.tabs,
      activeKey = props.activeKey,
      activeTabRef = props.activeTabRef,
      onTabClick = props.onTabClick,
      onTabKeyDown = props.onTabKeyDown;
  var tabNavClassName = "slds-tabs_".concat(type, "__nav");
  return /*#__PURE__*/_react["default"].createElement("ul", {
    className: tabNavClassName,
    role: "tablist"
  }, _react["default"].Children.map(tabs, function (tab) {
    return /*#__PURE__*/_react["default"].createElement(TabItem, (0, _extends2["default"])({}, tab.props, {
      type: type,
      activeKey: activeKey,
      activeTabRef: activeTabRef,
      onTabClick: onTabClick,
      onTabKeyDown: onTabKeyDown
    }));
  }));
};

/**
 *
 */
var Tab = function Tab(props) {
  var className = props.className,
      eventKey = props.eventKey,
      activeKey = props.activeKey,
      children = props.children;
  return /*#__PURE__*/_react["default"].createElement(TabContent, {
    className: className,
    active: eventKey === activeKey
  }, children);
};

exports.Tab = Tab;

/**
 *
 */
var Tabs = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Tabs, _Component);

  var _super = _createSuper(Tabs);

  function Tabs(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Tabs);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "activeTab", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onTabClick", function (tabKey) {
      if (_this.props.onSelect) {
        _this.props.onSelect(tabKey);
      } // Uncontrolled


      _this.setState({
        activeKey: tabKey,
        focusTab: true
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onTabKeyDown", function (tabKey, e) {
      if (e.keyCode === 37 || e.keyCode === 39) {
        // left/right cursor key
        var idx = 0;
        var tabKeys = [];

        _react["default"].Children.forEach(_this.props.children, function (tab, i) {
          tabKeys.push(tab.props.eventKey);

          if (tabKey === tab.props.eventKey) {
            idx = i;
          }
        });

        var dir = e.keyCode === 37 ? -1 : 1;
        var activeIdx = (idx + dir + tabKeys.length) % tabKeys.length;
        var activeKey = tabKeys[activeIdx];

        _this.onTabClick(activeKey);

        e.preventDefault();
        e.stopPropagation();
      }
    });
    _this.state = {};
    (0, _util.registerStyle)('tab-menu', [['.slds-tabs__item.react-slds-tab-with-menu', '{ position: relative !important; overflow: visible !important; }'], ['.slds-tabs__item.react-slds-tab-with-menu > .react-slds-tab-item-content', '{ overflow: hidden }'], ['.slds-tabs__item.react-slds-tab-with-menu > .react-slds-tab-item-content > a', '{ padding-right: 2rem; }'], ['.react-slds-tab-menu', '{ position: absolute; top: 0; right: 0; }'], ['.react-slds-tab-menu button', '{ height: 2.5rem; line-height: 2rem; width: 2rem; visibility: hidden; justify-content: center }'], ['.slds-tabs__item.slds-active .react-slds-tab-menu button', '.slds-tabs__item:hover .react-slds-tab-menu button', '.slds-tabs__item .react-slds-tab-menu button:focus', '{ visibility: visible }']]);
    return _this;
  }

  (0, _createClass2["default"])(Tabs, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (this.state.focusTab) {
        var el = this.activeTab;

        if (el) {
          el.focus();
        }
        /* eslint-disable react/no-did-update-set-state */


        this.setState({
          focusTab: false
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children;
      var type = this.props.type === 'scoped' ? 'scoped' : 'default';
      var tabsClassNames = (0, _classnames["default"])(className, "slds-tabs_".concat(type));
      var activeKey = typeof this.props.activeKey !== 'undefined' ? this.props.activeKey : typeof this.state.activeKey !== 'undefined' ? this.state.activeKey : this.props.defaultActiveKey;
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: tabsClassNames
      }, /*#__PURE__*/_react["default"].createElement(TabNav, {
        type: type,
        activeKey: activeKey,
        activeTabRef: function activeTabRef(node) {
          _this2.activeTab = node;
        },
        tabs: children,
        onTabClick: this.onTabClick,
        onTabKeyDown: this.onTabKeyDown
      }), _react["default"].Children.map(children, function (tab) {
        return /*#__PURE__*/_react["default"].cloneElement(tab, {
          activeKey: activeKey
        });
      }));
    }
  }]);
  return Tabs;
}(_react.Component);

exports.Tabs = Tabs;
