// react-lightning-design-system/lib/scripts/Toggle.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Toggle = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _FormElement = require("./FormElement");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Toggle = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Toggle, _Component);

  var _super = _createSuper(Toggle);

  function Toggle() {
    var _this;

    (0, _classCallCheck2["default"])(this, Toggle);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "node", null);
    return _this;
  }

  (0, _createClass2["default"])(Toggle, [{
    key: "renderToggle",
    value: function renderToggle(_ref) {
      var className = _ref.className,
          label = _ref.label,
          props = (0, _objectWithoutProperties2["default"])(_ref, ["className", "label"]);
      var toggleClassNames = (0, _classnames["default"])(className, 'slds-checkbox_toggle slds-grid');
      return /*#__PURE__*/_react["default"].createElement("label", {
        className: toggleClassNames
      }, /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-form-element__label slds-m-bottom_none"
      }, label), /*#__PURE__*/_react["default"].createElement("input", (0, _extends2["default"])({
        name: "checkbox",
        type: "checkbox",
        "aria-describedby": "toggle-desc"
      }, props)), /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-checkbox_faux_container",
        "aria-live": "assertive"
      }, /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-checkbox_faux"
      }), /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-checkbox_on"
      }, "Enabled"), /*#__PURE__*/_react["default"].createElement("span", {
        className: "slds-checkbox_off"
      }, "Disabled")));
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          required = _this$props.required,
          error = _this$props.error,
          totalCols = _this$props.totalCols,
          cols = _this$props.cols,
          props = (0, _objectWithoutProperties2["default"])(_this$props, ["required", "error", "totalCols", "cols"]);
      var formElemProps = {
        required: required,
        error: error,
        totalCols: totalCols,
        cols: cols
      };
      return /*#__PURE__*/_react["default"].createElement(_FormElement.FormElement, (0, _extends2["default"])({
        formElementRef: function formElementRef(node) {
          return _this2.node = node;
        }
      }, formElemProps), this.renderToggle(props));
    }
  }]);
  return Toggle;
}(_react.Component);

exports.Toggle = Toggle;
